<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Summarizing mutation status as a feature matrix • GAMBLR.open</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="feature_matrix_files/libs/quarto-html/popper.min.js"></script><script src="feature_matrix_files/libs/quarto-html/tippy.umd.min.js"></script><link href="feature_matrix_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="feature_matrix_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing mutation status as a feature matrix">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.open</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-open.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/glossary.html">GAMBLR terms and concepts</a></li>
    <li><a class="dropdown-item" href="../articles/mutation_summary.html">Exploring simple somatic mutations</a></li>
    <li><a class="dropdown-item" href="../articles/feature_matrix.html">Summarizing mutation status as a feature matrix</a></li>
    <li><a class="dropdown-item" href="../articles/genetic_classification.html">Genetic classification of tumours</a></li>
    <li><a class="dropdown-item" href="../articles/bundled_data.html">GAMBLR.data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Summarizing mutation status as a feature matrix</h1>





      <div class="d-none name"><code></code></div>
    </div>






    <section class="section level2"><h2 id="tutorial-assemble-feature-matrix">Tutorial: assemble feature matrix<a class="anchor" aria-label="anchor" href="#tutorial-assemble-feature-matrix"></a>
</h2>
<p>In 2023, the Morin Lab developed a classifier of Follicular Lymphoma (FL) predictive of histological transformation to more aggressive DLBCL. This study was published in <a href="https://doi.org/10.1182/blood.2022018719" class="external-link">Blood (2023).</a> How was the binary feature matrix assembled for that machine learning model?</p>
<p>In this quick tour we will show how GAMBLR.data resources and functionality of GAMBLR.open can help you to generate such matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
</div>
<p>This tutorial explores how to obtain metadata for the samples, simple somatic mutations in maf format, and auto-magically transform it into a binary matrix of features.</p>
<p>To simplify this tutorial, we will only include a small subset of features (not the whole set as was used in the original paper), but this example will be able to illustrate the functionality and highlight the main steps of the process.</p>
<section class="level2"><h3 id="obtain-metadata">Obtain metadata<a class="anchor" aria-label="anchor" href="#obtain-metadata"></a>
</h3>
<p>In the <a href="./getting_started.html">previous tutorial</a>, we have already explored the function <code><a href="../reference/get_gambl_metadata.html">get_gambl_metadata()</a></code>. We will use it to retreive the metadata:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">study</span> <span class="op">==</span> <span class="st">"FL_Dreval"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<p>We can now see that the samples from the FL study are in the cohort <code>FL_Dreval</code>. Let’s explore these samples more:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">pathology</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
COMFL DLBCL    FL
   21   209   213 </code></pre>
</div>
</div>
<p>For illustration purposes, let’s take 10 samples from the FL study: 5 FL and 5 DLBCL:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only filter for the samples from FL study</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"DLBCL"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_tail</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="va">ungroup</span></span>
<span></span>
<span><span class="co"># How does our metadata looks like now?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [10 × 30] (S3: tbl_df/tbl/data.frame)
 $ age_group           : chr [1:10] "Other" "Other" "Other" "Other" ...
 $ bam_available       : logi [1:10] TRUE TRUE TRUE TRUE TRUE TRUE ...
 $ cohort              : chr [1:10] "DLBCL_ICGC" "DLBCL_ICGC" "DLBCL_ICGC" "DLBCL_ICGC" ...
 $ compression         : chr [1:10] "cram" "cram" "cram" "cram" ...
 $ COO_consensus       : chr [1:10] "ABC" "GCB" NA "ABC" ...
 $ DHITsig_consensus   : chr [1:10] "DHITsigNeg" "DHITsigNeg" "NA" "DHITsigNeg" ...
 $ EBV_status_inf      : chr [1:10] NA NA NA NA ...
 $ ffpe_or_frozen      : chr [1:10] "frozen" "frozen" "frozen" "frozen" ...
 $ fl_grade            : chr [1:10] NA NA NA NA ...
 $ genetic_subgroup    : chr [1:10] "dFL" "dFL" "dFL" "dFL" ...
 $ genome_build        : chr [1:10] "hs37d5" "hs37d5" "hs37d5" "hs37d5" ...
 $ hiv_status          : chr [1:10] NA NA NA NA ...
 $ lymphgen            : chr [1:10] "Other" "Other" "BN2" "MCD-COMP" ...
 $ lymphgen_cnv_noA53  : chr [1:10] "Other" "Other" "BN2" "BN2/MCD" ...
 $ lymphgen_no_cnv     : chr [1:10] "MCD" "Other" "BN2" "BN2/MCD" ...
 $ lymphgen_with_cnv   : chr [1:10] "A53" "A53" "BN2" "BN2/MCD" ...
 $ lymphgen_wright     : chr [1:10] NA NA NA NA ...
 $ molecular_BL        : chr [1:10] "non-BL" "non-BL" NA "non-BL" ...
 $ normal_sample_id    : chr [1:10] "SP59410" "SP59446" "SP59450" "SP59454" ...
 $ pairing_status      : chr [1:10] "matched" "matched" "matched" "matched" ...
 $ pathology           : chr [1:10] "DLBCL" "DLBCL" "DLBCL" "DLBCL" ...
 $ pathology_rank      : num [1:10] 19 19 19 19 19 15 15 15 15 15
 $ patient_id          : chr [1:10] "DO27833" "DO27851" "DO27853" "DO27855" ...
 $ reference_PMID      : num [1:10] 37084389 37084389 37084389 37084389 37084389 ...
 $ sample_id           : chr [1:10] "SP59412" "SP59448" "SP59452" "SP59456" ...
 $ seq_type            : chr [1:10] "genome" "genome" "genome" "genome" ...
 $ sex                 : chr [1:10] "M" "F" "M" "F" ...
 $ study               : chr [1:10] "FL_Dreval" "FL_Dreval" "FL_Dreval" "FL_Dreval" ...
 $ time_point          : chr [1:10] "A" "A" "A" "A" ...
 $ Tumor_Sample_Barcode: chr [1:10] "SP59412" "SP59448" "SP59452" "SP59456" ...</code></pre>
</div>
</div>
<p>Our subsetting worked and we can now proceed with matrix assembly.</p>
</section><section class="level2"><h3 id="generate-feature-matrix">Generate feature matrix<a class="anchor" aria-label="anchor" href="#generate-feature-matrix"></a>
</h3>
<p>We will now use the metadata from the previous step to assemble the binary feature matrix.</p>
<section class="level3"><h4 id="return-the-simple-somatic-mutations">Return the simple somatic mutations<a class="anchor" aria-label="anchor" href="#return-the-simple-somatic-mutations"></a>
</h4>
<p>First, lets return the data frame with simple somatic mutations in maf format and store it in a variable. Technically this step is not strictly necessary, as each function used below will be able to retreive it for you when the maf data is not provided, but we will advocate for good practice here and have the maf data stored in a designated variable. As we will be utilizing both coding and non-coding mutations, we will take advantage of returning the variants per sample (without necessarily restricting to coding-only variants).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Obtain simple somatic mutations</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ssm_by_samples.html">get_ssm_by_samples</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>Wow, that was super easy and blazing fast! Did it even work?? Let’s confirm:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># How many samples are present in maf?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">maf</span><span class="op">$</span><span class="va">Tumor_Sample_Barcode</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Are all these samples the ones we are interested in and requested with metadata?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">maf</span><span class="op">$</span><span class="va">Tumor_Sample_Barcode</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">Tumor_Sample_Barcode</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># What are the mutations in the maf? Are they just coding?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">maf</span><span class="op">$</span><span class="va">Variant_Classification</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
               3'Flank                  3'UTR                5'Flank
                    29                     20                    127
                 5'UTR        Frame_Shift_Del        Frame_Shift_Ins
                    61                      7                      1
                   IGR           In_Frame_Del                 Intron
                    46                      2                   1502
     Missense_Mutation      Nonsense_Mutation                    RNA
                    78                      9                     15
                Silent          Splice_Region            Splice_Site
                    29                      6                      8
Translation_Start_Site
                     1 </code></pre>
</div>
</div>
<p>We can see from the above outputs that we got somatic mutations for all requested samples and the maf contains both coding and non-coding variants. We can now proceed to the next steps.</p>
</section><section class="level3"><h4 id="coding-mutations-as-features">Coding mutations as features<a class="anchor" aria-label="anchor" href="#coding-mutations-as-features"></a>
</h4>
<p>The feature matrix in the FL study contained coding mutation status at selected genes denoted as 1 when the mutation was present and 0 when there was no mutation. In addition to that, mutation hotspots at some genes were also taken into account. To complicate things more, annotation of hotspots was performed differently depending on the specific gene and mutational effect. Specifically, the <em>CREBBP</em> missemse mutations were considered hotspots when they occured in the KAT domain, but not outside of it. The mutations in <em>FOXO1</em> were considered hotspots when the AA change was at M1.</p>
<p>GAMBLR.data provides a one-stop shop to achieve this level of details out-of-the-box in a simple and convenient way, and is directly used by the <code>get_coding_ssm_status</code> function. This function has a logical arguments <code>include_hotspots</code> to separate regular mutations from the ones occurring at hotspots, and <code>review_hotspots</code>, which will handle the specific cases we described above in an automated way. Both of these arguments are <code>TRUE</code> by default, so you do not need to toggle them separately, but in this example we will specify them explicitly just to illustrate that this will happen during the function call. To keep the example clean and concise, we will also annotate the mutation status only for a few of selected genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify genes to annotate</span></span>
<span><span class="va">our_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"CREBBP"</span>, <span class="st">"MYD88"</span>, <span class="st">"RRAGC"</span>,</span>
<span>    <span class="st">"PIM1"</span>, <span class="st">"BCL2"</span>, <span class="st">"BCL6"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate binary matrix</span></span>
<span><span class="va">coding_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_coding_ssm_status.html">get_coding_ssm_status</a></span><span class="op">(</span></span>
<span>    gene_symbols <span class="op">=</span> <span class="va">our_genes</span>,</span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_data <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    include_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    review_hotspots <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">coding_matrix</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   sample_id PIM1 CREBBP RRAGC MYD88 BCL6 BCL2 CREBBPHOTSPOT MYD88HOTSPOT
1    SP59412    1      1     0     0    0    0             0            0
2    SP59448    0      0     0     0    0    0             0            0
3    SP59452    1      0     0     1    1    0             0            0
4    SP59456    0      0     0     0    0    0             0            0
5    SP59460    0      0     0     0    1    1             0            1
6    SP59420    0      1     0     0    0    0             0            0
7    SP59424    0      0     0     0    0    0             1            0
8    SP59432    0      0     1     0    0    0             0            0
9    SP59436    0      1     0     0    0    0             0            0
10   SP59464    0      0     1     0    0    0             0            0</code></pre>
</div>
</div>
<p>We can see that in this example there is only one sample with hotspot mutation in <em>CREBBP</em>, SP59424 (annotated as <code>1</code> in the column <code>CREBBPHOTSPOT</code>). Let’s sanity check this annotation for illustration purposes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crebbp_hotspot_mutation</span> <span class="op">&lt;-</span> <span class="va">maf</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">Tumor_Sample_Barcode</span> <span class="op">==</span> <span class="st">"SP59424"</span>,</span>
<span>        <span class="va">Hugo_Symbol</span> <span class="op">==</span> <span class="st">"CREBBP"</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Chromosome</span>, <span class="va">Start_Position</span>, <span class="va">Variant_Classification</span><span class="op">)</span></span>
<span></span>
<span><span class="va">crebbp_hotspot_mutation</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>genomic_data Object
Genome Build: grch37
Showing first 10 rows:
  Chromosome Start_Position Variant_Classification
1         16        3788650      Missense_Mutation</code></pre>
</div>
</div>
<p>This mutation falls within KAT domain and is a missense variant, so it is indeed correct to be annotated as hotspot. How can we check it does fall within KAT domain? This is easy with GAMBLR.data!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># GAMBLR.data has the regions to be considered as hotspots</span></span>
<span><span class="va">hotspot_regions_grch37</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               chrom     start       end
CREBBP            16   3785000   3791000
EZH2               7 148508764 148506238
NOTCH1             9 139391455 139391455
NOTCH2             1 120459150 120459150
CD79B_trunc       17  62007172  62007172
CD79B_NONtrunc    17  62006800  62006800</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Now check that CREBBP mutations</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/between.html" class="external-link">between</a></span><span class="op">(</span></span>
<span>    <span class="va">crebbp_hotspot_mutation</span><span class="op">$</span><span class="va">Start_Position</span>,</span>
<span>    <span class="va">hotspot_regions_grch37</span><span class="op">[</span><span class="st">"CREBBP"</span>, <span class="st">"start"</span><span class="op">]</span>,</span>
<span>    <span class="va">hotspot_regions_grch37</span><span class="op">[</span><span class="st">"CREBBP"</span>, <span class="st">"end"</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>We have now generated matrix of coding mutations and SSM hotspots in a binary format and are ready to proceed to the next step.</p>
</section><section class="level3"><h4 id="ashm-mutations-as-features">aSHM mutations as features<a class="anchor" aria-label="anchor" href="#ashm-mutations-as-features"></a>
</h4>
<p>The FL study also annotated the non-coding mutations at selected aSHM targets as features of binary matrix. Again, all the necessarily means to do it in a simple step are provided by GAMBLR.data. The regions targeted by aSHM <a href="https://morinlab.github.io/GAMBLR.data/resources/bundled_data.html#ashm-regions" class="external-link">are available</a> in the GAMBLR.data. Since they are always complemented with new regions once they are identified, the latest and most comprehensive version is always available by referring to <code>{projection}_ashm_regions</code>. However, this use case is a perfect example to demonstarate how to operate on versioned iterations of the aSHM target list, as the list has been updated since the time the study was published and at the time of publication version 0.2 was used. We can refer to it directly by the version number:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions_bed</span> <span class="op">&lt;-</span> <span class="va">somatic_hypermutation_locations_GRCh37_v0.2</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">regions_bed</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  chr_name hg19_start  hg19_end    gene     region regulatory_comment
1     chr2   96808901  96811913   DUSP2   intron-1           enhancer
2    chr17   56407732  56410140 TSPOAP1 intergenic           enhancer
3    chr11  128339774 128345731    ETS1    introns           enhancer
4    chr11  128388492 128394163    ETS1      TSS-2    active_promoter
5     chr6   31548325  31550717     LTB   intron-1           enhancer
6     chr3   32020518  32024930 OSBPL10      TSS-1    active_promoter</code></pre>
</div>
</div>
<p>We will perform some simple modifications to it to make our experience better and only select few regions for illustrative purposes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">our_regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"BCL6-TSS"</span>,</span>
<span>    <span class="st">"BCL7A-TSS"</span>,</span>
<span>    <span class="st">"RHOH-TSS"</span>,</span>
<span>    <span class="st">"ZFP36L1-TSS"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">regions_bed</span> <span class="op">&lt;-</span> <span class="fu">create_bed_data</span><span class="op">(</span></span>
<span>    <span class="va">regions_bed</span>,</span>
<span>    fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span>
<span>    concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>,<span class="st">"region"</span><span class="op">)</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>    genome_build <span class="op">=</span> <span class="st">"grch37"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">regions_bed</span> <span class="op">&lt;-</span> <span class="va">regions_bed</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">our_regions</span><span class="op">)</span></span>
<span></span>
<span><span class="va">regions_bed</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>genomic_data Object
Genome Build: grch37
Showing first 10 rows:
  chrom     start       end        name region regulatory_comment
1     3 187458526 187464632    BCL6-TSS    TSS               &lt;NA&gt;
2    12 122456912 122464036   BCL7A-TSS    TSS    poised_promoter
3    14  69257848  69259739 ZFP36L1-TSS    TSS    active_promoter
4     4  40193105  40204231    RHOH-TSS    TSS    active_promoter</code></pre>
</div>
</div>
<p>Now we can see whether or not there are any mutations within these regions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ashm_matrix</span> <span class="op">&lt;-</span> <span class="fu">cool_overlaps</span><span class="op">(</span></span>
<span>    <span class="va">maf</span>,</span>
<span>    <span class="va">regions_bed</span>,</span>
<span>    columns2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrom"</span>, <span class="st">"start"</span>, <span class="st">"end"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Tumor_Sample_Barcode</span>, <span class="va">name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span></span>
<span>        id_cols <span class="op">=</span> <span class="va">Tumor_Sample_Barcode</span>,</span>
<span>        names_from <span class="op">=</span> <span class="va">name</span>,</span>
<span>        values_from <span class="op">=</span> <span class="va">n</span>,</span>
<span>        values_fill <span class="op">=</span> <span class="fl">0</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">"Tumor_Sample_Barcode"</span><span class="op">)</span></span>
<span><span class="va">ashm_matrix</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        BCL6-TSS ZFP36L1-TSS BCL7A-TSS RHOH-TSS
SP59412        4           1         0        0
SP59420        1           0         1        0
SP59424        0           0         0        1
SP59436        1           0         0        0
SP59448        6           4        14        3
SP59452       18           2        12       28
SP59456        2           6         0        9
SP59460        1           1         1        2
SP59464        1           0         0        0</code></pre>
</div>
</div>
<p>We now calculated the number of mutations in each region for each sample. The original study used pathology-specific values per region to convert these counts to binary, but here we will use an arbitrary cutoff of 5 to binarize these counts. We will also convert rownames to column so it will be easier for us to unite all matrices into single one at the later steps:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ashm_matrix</span><span class="op">[</span><span class="va">ashm_matrix</span> <span class="op">&lt;=</span> <span class="fl">5</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="va">ashm_matrix</span><span class="op">[</span><span class="va">ashm_matrix</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="va">ashm_matrix</span> <span class="op">&lt;-</span> <span class="va">ashm_matrix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ashm_matrix</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  sample_id BCL6-TSS ZFP36L1-TSS BCL7A-TSS RHOH-TSS
1   SP59412        0           0         0        0
2   SP59420        0           0         0        0
3   SP59424        0           0         0        0
4   SP59436        0           0         0        0
5   SP59448        1           0         1        0
6   SP59452        1           0         1        1
7   SP59456        0           1         0        1
8   SP59460        0           0         0        0
9   SP59464        0           0         0        0</code></pre>
</div>
</div>
<p>We have now generated matrix of non-coding mutations in a binary format and are ready to unite all matrices together.</p>
</section><section class="level3"><h4 id="unite-into-full-feature-matrix">Unite into full feature matrix<a class="anchor" aria-label="anchor" href="#unite-into-full-feature-matrix"></a>
</h4>
<p>We can now combine both coding and non-coding mutations into single matrix:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>    <span class="va">coding_matrix</span>,</span>
<span>    <span class="va">ashm_matrix</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span></span>
<span>            <span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span>, <span class="op">~</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/replace_na.html" class="external-link">replace_na</a></span><span class="op">(</span><span class="va">.x</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">feature_matrix</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   sample_id PIM1 CREBBP RRAGC MYD88 BCL6 BCL2 CREBBPHOTSPOT MYD88HOTSPOT
1    SP59412    1      1     0     0    0    0             0            0
2    SP59448    0      0     0     0    0    0             0            0
3    SP59452    1      0     0     1    1    0             0            0
4    SP59456    0      0     0     0    0    0             0            0
5    SP59460    0      0     0     0    1    1             0            1
6    SP59420    0      1     0     0    0    0             0            0
7    SP59424    0      0     0     0    0    0             1            0
8    SP59432    0      0     1     0    0    0             0            0
9    SP59436    0      1     0     0    0    0             0            0
10   SP59464    0      0     1     0    0    0             0            0
   BCL6-TSS ZFP36L1-TSS BCL7A-TSS RHOH-TSS
1         0           0         0        0
2         1           0         1        0
3         1           0         1        1
4         0           1         0        1
5         0           0         0        0
6         0           0         0        0
7         0           0         0        0
8         0           0         0        0
9         0           0         0        0
10        0           0         0        0</code></pre>
</div>
</div>
<p>That’s it!</p>
<p><em>Happy GAMBLing!</em></p>
<pre><code>  /$$$$$$     /$$$$$$    /$$      /$$   /$$$$$$$    /$$        .:::::::
 /$$__  $$   /$$__  $$  | $$$    /$$$  | $$__  $$  | $$        .::    .::
| $$  \__/  | $$  \ $$  | $$$$  /$$$$  | $$  \ $$  | $$        .::    .::
| $$ /$$$$  | $$$$$$$$  | $$ $$/$$ $$  | $$$$$$$   | $$   &lt;-   .: .::
| $$|_  $$  | $$__  $$  | $$  $$$| $$  | $$__  $$  | $$        .::  .::
| $$  \ $$  | $$  | $$  | $$\  $ | $$  | $$  \ $$  | $$        .::    .::
|  $$$$$$/  | $$  | $$  | $$ \/  | $$  | $$$$$$$/  | $$$$$$$$  .::      .::
 \______/   |__/  |__/  |__/     |__/  |_______/   |________/
 ~GENOMIC~~~~~~~~~~~~~OF~~~~~~~~~~~~~~~~~B-CELL~~~~~~~~~~~~~~~~~~IN~~~~~~
 ~~~~~~~~~~~~ANALYSIS~~~~~~MATURE~~~~~~~~~~~~~~~~~~~LYMPHOMAS~~~~~~~~~~R~ </code></pre>
</section></section></section><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'light-border',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config);
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostia Dreval.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
