<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Applying a DLBCLone model to new data • GAMBLR.predict</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="validation_files/libs/quarto-html/popper.min.js"></script><script src="validation_files/libs/quarto-html/tippy.umd.min.js"></script><link href="validation_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="validation_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Applying a DLBCLone model to new data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.predict</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/index.html">DLBCLone: Overview</a></li>
    <li><a class="dropdown-item" href="../articles/umap.html">Exploring and selecting features</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_optimize_params.html">Training a new DLBCLone model</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_stored.html">DLBCLone Models</a></li>
    <li><a class="dropdown-item" href="../articles/predict_samples.html">Classifying samples with DLBCLone</a></li>
    <li><a class="dropdown-item" href="../articles/validation.html">Applying a DLBCLone model to new data</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_KNN.html">K-Nearest Neighbors without UMAP</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_mixture_model.html">Gaussian Mixture Model</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Applying a DLBCLone model to new data</h1>





      <div class="d-none name"><code></code></div>
    </div>






    <section class="level2"><h2 id="preamble">Preamble<a class="anchor" aria-label="anchor" href="#preamble"></a>
</h2>
<p>This guide will demonstrate the steps for training a custom DLBCLone model and applying it to classify new samples.</p>
</section><section class="level2"><h2 id="preparing-your-test-data">Preparing your <em>test</em> data<a class="anchor" aria-label="anchor" href="#preparing-your-test-data"></a>
</h2>
<p>To use a trained DLBCLone model on your data, you need (at a minimum) either a MAF file or a mutation status matrix for the genes (features) that were used to generate the model. Here, you will learn how to generate a mutation status matrix using one of our published cohorts, which was sequenced with the LySeqST panel. If you have translocation status for BCL2, BCL6 etc and want to include that in the model, you will also need a second table that includes the translocation status for each sample. We generally refer to this as the metadata table. Together, your MAF and metadata files will be used to assemble the feature matrix that will be used for UMAP and KNN approximation on your samples.</p>
<section class="level3"><h3 id="sv-status">SV status<a class="anchor" aria-label="anchor" href="#sv-status"></a>
</h3>
<p>For this example, we will extract the BCL2 and BCL6 status directly from the output of LymphGen. We will also use the LymphGen classification to evaluate our accuracy. Importantly, you do not need a LymphGen result from your <em>test</em> data but it can be useful to sanity check your UMAP and your model performance. The MAF and metadata for these samples (in the form of a LymphGen output) are both bundled with this package. The code below loads these files and restructures a few columns to make them compatible with downstream functions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.utils</span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.predict</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lymphgen_test</span> <span class="op">=</span> <span class="st">"extdata/lyseq_validation/LySeqST_only_unaugmented_vaf.lymphgen_calls.no_cnvs.with_sv.no_A53.tsv"</span></span>
<span><span class="va">valid_metadata</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="va">lymphgen_test</span>,package <span class="op">=</span> <span class="st">"GAMBLR.predict"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span></span>
<span>    sample_id <span class="op">=</span> <span class="va">Sample.Name</span>,</span>
<span>    lymphgen <span class="op">=</span> <span class="va">Subtype.Prediction</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    BCL2.Translocation <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">BCL2.Translocation</span> <span class="op">==</span> <span class="st">"yes"</span>, <span class="st">"POS"</span>, <span class="va">BCL2.Translocation</span><span class="op">)</span>,</span>
<span>    BCL2.Translocation <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">BCL2.Translocation</span> <span class="op">==</span> <span class="st">"no"</span>, <span class="st">"NEG"</span>, <span class="va">BCL2.Translocation</span><span class="op">)</span>,</span>
<span>    BCL2.Translocation <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">BCL2.Translocation</span> <span class="op">==</span> <span class="st">"Not Available"</span>, <span class="st">"NA"</span>, <span class="va">BCL2.Translocation</span><span class="op">)</span>,</span>
<span>    BCL6.Translocation <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">BCL6.Translocation</span> <span class="op">==</span> <span class="st">"yes"</span>, <span class="st">"POS"</span>, <span class="va">BCL6.Translocation</span><span class="op">)</span>,</span>
<span>    BCL6.Translocation <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">BCL6.Translocation</span> <span class="op">==</span> <span class="st">"no"</span>, <span class="st">"NEG"</span>, <span class="va">BCL6.Translocation</span><span class="op">)</span>,</span>
<span>    BCL6.Translocation <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">BCL6.Translocation</span> <span class="op">==</span> <span class="st">"Not Available"</span>, <span class="st">"NA"</span>, <span class="va">BCL6.Translocation</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">maf_test</span> <span class="op">=</span> <span class="st">"extdata/lyseq_validation/LySeqST_merged_unaugmented_vaf.maf"</span></span>
<span><span class="va">valid_maf_vaf_dir</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="va">maf_test</span>,package <span class="op">=</span> <span class="st">"GAMBLR.predict"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>  </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    sample_id <span class="op">=</span> <span class="va">Tumor_Sample_Barcode</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#convert to S3 object for compatability</span></span>
<span><span class="va">valid_maf_vaf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.utils/man/create_maf_data.html" class="external-link">create_maf_data</a></span><span class="op">(</span></span>
<span>    maf_df <span class="op">=</span> <span class="va">valid_maf_vaf_dir</span>,</span>
<span>    genome_build <span class="op">=</span> <span class="st">"grch37"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Chromosome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">Chromosome</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level3"><h3 id="synonymous-mutations">Synonymous mutations<a class="anchor" aria-label="anchor" href="#synonymous-mutations"></a>
</h3>
<p>If the model you’re using used base-3 encoding (0,1,2), that means synonymous mutations are considered for some genes. You will have to know up-front (or find out) which genes this applies to, so you can encode your data consistently. If your MAF file doesn’t contain any synonyous mutations, be sure to use a model that was trained only on coding mutations.</p>
</section></section><section class="level2"><h2 id="metadata">metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">valid_metadata</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<table class="table caption-top" style="width:100%;">
<colgroup>
<col style="width: 3%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 12%">
<col style="width: 15%">
<col style="width: 8%">
<col style="width: 3%">
<col style="width: 8%">
<col style="width: 1%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">Copy.Number</th>
<th style="text-align: left;">BCL2.Translocation</th>
<th style="text-align: left;">BCL6.Translocation</th>
<th style="text-align: left;">Model.Used</th>
<th style="text-align: right;">Confidence.BN2</th>
<th style="text-align: right;">Confidence.EZB</th>
<th style="text-align: right;">Confidence.MCD</th>
<th style="text-align: right;">Confidence.N1</th>
<th style="text-align: right;">Confidence.ST2</th>
<th style="text-align: right;">BN2.Feature.Count</th>
<th style="text-align: right;">EZB.Feature.Count</th>
<th style="text-align: right;">MCD.Feature.Count</th>
<th style="text-align: right;">N1.Feature.Count</th>
<th style="text-align: right;">ST2.Feature.Count</th>
<th style="text-align: left;">BN2.Features</th>
<th style="text-align: left;">EZB.Features</th>
<th style="text-align: left;">MCD.Features</th>
<th style="text-align: left;">N1.Features</th>
<th style="text-align: left;">ST2.Features</th>
<th style="text-align: left;">lymphgen</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">00-14595_tumorA</td>
<td style="text-align: left;">Not Available</td>
<td style="text-align: left;">POS</td>
<td style="text-align: left;">POS</td>
<td style="text-align: left;">NOCGH</td>
<td style="text-align: right;">0.79</td>
<td style="text-align: right;">0.99</td>
<td style="text-align: right;">0.00</td>
<td style="text-align: right;">0.07</td>
<td style="text-align: right;">0.03</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">BCL6FusionCatcherORInHouse_FUSION,DTX1_MUTATION,BTG2_SubSyn</td>
<td style="text-align: left;">BCL2TransFISHMUTFusion_CompUp,EZH2_SubMUTATION,SOCS1_MUTATION</td>
<td style="text-align: left;">BTG2_SubSyn</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">SOCS1_MUTATION,ITPKB_MUTATION</td>
<td style="text-align: left;">EZB</td>
</tr>
<tr class="even">
<td style="text-align: left;">00-14595_tumorB</td>
<td style="text-align: left;">Not Available</td>
<td style="text-align: left;">POS</td>
<td style="text-align: left;">POS</td>
<td style="text-align: left;">NOCGH</td>
<td style="text-align: right;">0.79</td>
<td style="text-align: right;">0.99</td>
<td style="text-align: right;">0.00</td>
<td style="text-align: right;">0.07</td>
<td style="text-align: right;">0.68</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3</td>
<td style="text-align: left;">BCL6FusionCatcherORInHouse_FUSION,DTX1_MUTATION,BTG2_SubSyn</td>
<td style="text-align: left;">BCL2TransFISHMUTFusion_CompUp,EZH2_SubMUTATION,SOCS1_MUTATION</td>
<td style="text-align: left;">TBL1XR1_MUTATION,PIM2_SubSyn,BTG2_SubSyn</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">SGK1_TRUNC,SOCS1_MUTATION,ITPKB_MUTATION</td>
<td style="text-align: left;">EZB</td>
</tr>
<tr class="odd">
<td style="text-align: left;">00-14595_tumorC</td>
<td style="text-align: left;">Not Available</td>
<td style="text-align: left;">POS</td>
<td style="text-align: left;">POS</td>
<td style="text-align: left;">NOCGH</td>
<td style="text-align: right;">0.79</td>
<td style="text-align: right;">0.99</td>
<td style="text-align: right;">0.00</td>
<td style="text-align: right;">0.07</td>
<td style="text-align: right;">0.68</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3</td>
<td style="text-align: left;">BCL6FusionCatcherORInHouse_FUSION,DTX1_MUTATION,BTG2_SubSyn</td>
<td style="text-align: left;">BCL2TransFISHMUTFusion_CompUp,EZH2_SubMUTATION,KMT2D_TRUNC,SOCS1_MUTATION</td>
<td style="text-align: left;">TBL1XR1_MUTATION,PIM2_SubSyn,BTG2_SubSyn</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">SGK1_TRUNC,SOCS1_MUTATION,ITPKB_MUTATION</td>
<td style="text-align: left;">EZB</td>
</tr>
<tr class="even">
<td style="text-align: left;">00-14595_tumorD</td>
<td style="text-align: left;">Not Available</td>
<td style="text-align: left;">POS</td>
<td style="text-align: left;">NEG</td>
<td style="text-align: left;">NOCGH</td>
<td style="text-align: right;">0.17</td>
<td style="text-align: right;">0.98</td>
<td style="text-align: right;">0.00</td>
<td style="text-align: right;">0.07</td>
<td style="text-align: right;">0.00</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">DTX1_MUTATION,BTG2_SubSyn</td>
<td style="text-align: left;">BCL2TransFISHMUTFusion_CompUp,EZH2_SubMUTATION</td>
<td style="text-align: left;">PIM2_SubSyn,BTG2_SubSyn</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">EZB</td>
</tr>
<tr class="odd">
<td style="text-align: left;">00-15201_tumorA</td>
<td style="text-align: left;">Not Available</td>
<td style="text-align: left;">NEG</td>
<td style="text-align: left;">NEG</td>
<td style="text-align: left;">NOCGH</td>
<td style="text-align: right;">0.07</td>
<td style="text-align: right;">0.05</td>
<td style="text-align: right;">0.05</td>
<td style="text-align: right;">0.07</td>
<td style="text-align: right;">0.15</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3</td>
<td style="text-align: left;">TNFAIP3_TRUNC</td>
<td style="text-align: left;">KMT2D_TRUNC</td>
<td style="text-align: left;">PIM1_Synon,BTG1_SubSyn,HLA-A_MUTATION</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">CD83_Synon,DUSP2_MUTATION,WEE1_Synon</td>
<td style="text-align: left;">Other</td>
</tr>
<tr class="even">
<td style="text-align: left;">00-15201_tumorB</td>
<td style="text-align: left;">Not Available</td>
<td style="text-align: left;">NEG</td>
<td style="text-align: left;">NEG</td>
<td style="text-align: left;">NOCGH</td>
<td style="text-align: right;">0.18</td>
<td style="text-align: right;">0.11</td>
<td style="text-align: right;">0.01</td>
<td style="text-align: right;">0.99</td>
<td style="text-align: right;">0.00</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">TNFAIP3_TRUNC,KLF2_Synon</td>
<td style="text-align: left;">EP300_MUTATION</td>
<td style="text-align: left;">HLA-B_MUTATION,PIM2_SubSyn,HLA-A_MUTATION</td>
<td style="text-align: left;">NOTCH1_MUTATION</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">N1</td>
</tr>
</tbody>
</table>
</div>
</div>
</section><section class="level2"><h2 id="lyseq-features">lyseq features<a class="anchor" aria-label="anchor" href="#lyseq-features"></a>
</h2>
<p>The LySeqST sequencing panel includes the following genes. In order to train a classifier compatible with this panel, we need to subset the training data to ensure only these genes are included in training.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lyseq_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"TNFRSF14"</span>, <span class="st">"SPEN"</span>, <span class="st">"ID3"</span>, <span class="st">"ARID1A"</span>, <span class="st">"RRAGC"</span>,</span>
<span>  <span class="st">"BCL10"</span>, <span class="st">"CD58"</span>, <span class="st">"NOTCH2"</span>, <span class="st">"FCGR2B"</span>, <span class="st">"FCRLA"</span>,</span>
<span>  <span class="st">"FAS"</span>, <span class="st">"CCND1"</span>, <span class="st">"BIRC3"</span>, <span class="st">"ATM"</span>, <span class="st">"KMT2D"</span>,</span>
<span>  <span class="st">"STAT6"</span>, <span class="st">"BTG1"</span>, <span class="st">"FOXO1"</span>, <span class="st">"B2M"</span>, <span class="st">"MAP2K1"</span>,</span>
<span>  <span class="st">"IDH2"</span>, <span class="st">"CHD2"</span>, <span class="st">"CREBBP"</span>, <span class="st">"SOCS1"</span>, <span class="st">"IL4R"</span>,</span>
<span>  <span class="st">"PLCG2"</span>, <span class="st">"TP53"</span>, <span class="st">"STAT5B"</span>, <span class="st">"STAT3"</span>, <span class="st">"CD79B"</span>,</span>
<span>  <span class="st">"GNA13"</span>, <span class="st">"BCL2"</span>, <span class="st">"TCF3"</span>, <span class="st">"S1PR2"</span>, <span class="st">"JAK3"</span>,</span>
<span>  <span class="st">"MEF2B"</span>, <span class="st">"DNMT3A"</span>, <span class="st">"XPO1"</span>, <span class="st">"CXCR4"</span>, <span class="st">"SF3B1"</span>,</span>
<span>  <span class="st">"PTPN1"</span>, <span class="st">"XBP1"</span>, <span class="st">"EP300"</span>, <span class="st">"MYD88"</span>, <span class="st">"SETD2"</span>,</span>
<span>  <span class="st">"RHOA"</span>, <span class="st">"NFKBIZ"</span>, <span class="st">"TBL1XR1"</span>, <span class="st">"KLHL6"</span>, <span class="st">"TET2"</span>,</span>
<span>  <span class="st">"PIM1"</span>, <span class="st">"CCND3"</span>, <span class="st">"TMEM30A"</span>, <span class="st">"PRDM1"</span>, <span class="st">"SGK1"</span>,</span>
<span>  <span class="st">"TNFAIP3"</span>, <span class="st">"CARD11"</span>, <span class="st">"POT1"</span>, <span class="st">"BRAF"</span>, <span class="st">"EZH2"</span>,</span>
<span>  <span class="st">"UBR5"</span>, <span class="st">"MYC"</span>, <span class="st">"NOTCH1"</span>, <span class="st">"TRAF2"</span>, <span class="st">"P2RY8"</span>,</span>
<span>  <span class="st">"BTK"</span>, <span class="st">"TP73"</span>, <span class="st">"NOL9"</span>, <span class="st">"SEMA4A"</span>, <span class="st">"PRRC2C"</span>,</span>
<span>  <span class="st">"BTG2"</span>, <span class="st">"ITPKB"</span>, <span class="st">"SEC24C"</span>, <span class="st">"EDRF1"</span>, <span class="st">"WEE1"</span>,</span>
<span>  <span class="st">"MPEG1"</span>, <span class="st">"ETS1"</span>, <span class="st">"DTX1"</span>, <span class="st">"SETD1B"</span>, <span class="st">"NFKBIA"</span>,</span>
<span>  <span class="st">"ZFP36L1"</span>, <span class="st">"CIITA"</span>, <span class="st">"MBTPS1"</span>, <span class="st">"IRF8"</span>, <span class="st">"ACTG1"</span>,</span>
<span>  <span class="st">"KLHL14"</span>, <span class="st">"MED16"</span>, <span class="st">"CD70"</span>, <span class="st">"JUNB"</span>, <span class="st">"KLF2"</span>,</span>
<span>  <span class="st">"CD79A"</span>, <span class="st">"DYSF"</span>, <span class="st">"DUSP2"</span>, <span class="st">"BCL2L1"</span>, <span class="st">"PRDM15"</span>,</span>
<span>  <span class="st">"RFTN1"</span>, <span class="st">"OSBPL10"</span>, <span class="st">"EIF4A2"</span>, <span class="st">"BCL6"</span>, <span class="st">"IRF4"</span>,</span>
<span>  <span class="st">"FOXC1"</span>, <span class="st">"CD83"</span>, </span>
<span>  <span class="st">"HLA-A"</span>, <span class="st">"HLA-B"</span>, <span class="st">"PRRC2A"</span>, <span class="st">"TBCC"</span>, <span class="st">"INTS1"</span>,</span>
<span>  <span class="st">"ACTB"</span>, <span class="st">"PIK3CG"</span>, <span class="st">"PRKDC"</span>, <span class="st">"TOX"</span>, <span class="st">"CDKN2A"</span>,</span>
<span>  <span class="st">"GRHPR"</span>, <span class="st">"DDX3X"</span>, <span class="st">"PIM2"</span>, <span class="st">"UBE2A"</span>, <span class="st">"ETV6"</span>,</span>
<span>  <span class="st">"MS4A1"</span>, <span class="st">"CD19"</span>, <span class="st">"HNRNPD"</span>, <span class="st">"NFKBIE"</span>, <span class="st">"TMSB4X"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="making-a-compatible-feature-matrix">Making a compatible feature matrix<a class="anchor" aria-label="anchor" href="#making-a-compatible-feature-matrix"></a>
</h2>
<p>You create a mutation feature matrix with <code>assemble_genetic_features</code>. This can include mutation status and structural variant status for BCL2, BCL6, and MYC. By default, synonymous mutations will be encoded as 1 and SV and coding mutations will be encoded as 2. The genes for which synonymous mutations are considered are specified separately. The genes we commonly use are shown in the example below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ashm_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ACTB"</span>,<span class="st">"ACTG1"</span>,<span class="st">"BCL2"</span>,<span class="st">"BCL6"</span>,<span class="st">"BIRC3"</span>,</span>
<span>     <span class="st">"BTG1"</span>,<span class="st">"BTG2"</span>,<span class="st">"CD83"</span>,<span class="st">"CIITA"</span>,<span class="st">"CXCR4"</span>,<span class="st">"DDX3X"</span>,    </span>
<span>     <span class="st">"DTX1"</span>,<span class="st">"DUSP2"</span>,<span class="st">"ETS1"</span>,<span class="st">"ETV6"</span>,<span class="st">"GRHPR"</span>,<span class="st">"ID3"</span>, </span>
<span>     <span class="st">"IL4R"</span>,<span class="st">"IRF4"</span>,<span class="st">"IRF8"</span>,<span class="st">"ITPKB"</span>,<span class="st">"KLF2"</span>,<span class="st">"KLHL6"</span>,</span>
<span>     <span class="st">"MEF2B"</span>,<span class="st">"MS4A1"</span>,<span class="st">"MYC"</span>,<span class="st">"NFKBIZ"</span>,<span class="st">"OSBPL10"</span>,<span class="st">"PIM1"</span>,</span>
<span>     <span class="st">"PIM2"</span>,<span class="st">"PTPN1"</span>,<span class="st">"RFTN1"</span>,<span class="st">"S1PR2"</span>,<span class="st">"SGK1"</span>,<span class="st">"SOCS1"</span>,</span>
<span>     <span class="st">"TMSB4X"</span>,<span class="st">"WEE1"</span>,<span class="st">"XBP1"</span>,<span class="st">"ZFP36L1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lyseq_valid</span> <span class="op">=</span> <span class="fu"><a href="../reference/assemble_genetic_features.html">assemble_genetic_features</a></span><span class="op">(</span></span>
<span>  these_samples_metadata <span class="op">=</span> <span class="va">valid_metadata</span>,</span>
<span>  sv_from_metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>BCL2 <span class="op">=</span> <span class="st">'BCL2.Translocation'</span>, BCL6 <span class="op">=</span> <span class="st">'BCL6.Translocation'</span><span class="op">)</span>,</span>
<span>  genes<span class="op">=</span><span class="va">lyseq_genes</span>,</span>
<span>  maf_with_synon <span class="op">=</span> <span class="va">valid_maf_vaf</span>,</span>
<span>  hotspot_genes <span class="op">=</span> <span class="st">"MYD88"</span>,</span>
<span>  synon_genes<span class="op">=</span><span class="va">ashm_genes</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="load-the-training-data">Load the training data<a class="anchor" aria-label="anchor" href="#load-the-training-data"></a>
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_full_status_train</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/all_full_status.tsv"</span>,package <span class="op">=</span> <span class="st">"GAMBLR.predict"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dlbcl_meta</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/dlbcl_meta_with_dlbclass.tsv"</span>,package <span class="op">=</span> <span class="st">"GAMBLR.predict"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dlbcl_meta_clean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">dlbcl_meta</span>,</span>
<span>                          <span class="va">lymphgen</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MCD"</span>,<span class="st">"EZB"</span>,<span class="st">"BN2"</span>,<span class="st">"N1"</span>,<span class="st">"ST2"</span>,<span class="st">"Other"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Before training, the mutation matrix from the bundled training data must be subset to exclude any features that are not part of your panel.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">all_full_status_train</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 202</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">lyseq_valid</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 117</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lyseq_train</span> <span class="op">&lt;-</span> <span class="va">all_full_status_train</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">lyseq_valid</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="create-training-data-umap">Create training data UMAP<a class="anchor" aria-label="anchor" href="#create-training-data-umap"></a>
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mu_lyseq_train</span> <span class="op">=</span> <span class="fu"><a href="../reference/make_and_annotate_umap.html">make_and_annotate_umap</a></span><span class="op">(</span></span>
<span>  <span class="va">lyseq_train</span>,</span>
<span>  <span class="va">dlbcl_meta_clean</span>,</span>
<span>  core_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    ST2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SGK1"</span>,<span class="st">"DUSP2"</span>,<span class="st">"TET2"</span>,<span class="st">"SOCS1"</span><span class="op">)</span>,</span>
<span>    N1<span class="op">=</span><span class="st">"NOTCH1"</span>,</span>
<span>    EZB<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EZH2"</span>,<span class="st">"BCL2_SV"</span><span class="op">)</span>,</span>
<span>    MCD<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MYD88HOTSPOT"</span>,<span class="st">"CD79B"</span>,<span class="st">"PIM1"</span><span class="op">)</span>,</span>
<span>    BN2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCL6_SV"</span>,<span class="st">"NOTCH2"</span>,<span class="st">"SPEN"</span>,<span class="st">"CD70"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<section class="level3"><h3 id="visualize-umap">Visualize UMAP<a class="anchor" aria-label="anchor" href="#visualize-umap"></a>
</h3>
<p>Visually confirm the separation of the training samples based on their LymphGen class labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_umap_scatterplot.html">make_umap_scatterplot</a></span><span class="op">(</span><span class="va">mu_lyseq_train</span><span class="op">$</span><span class="va">df</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="validation_files/figure-html/unnamed-chunk-2-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section class="level2"><h2 id="train-the-model">Train the model<a class="anchor" aria-label="anchor" href="#train-the-model"></a>
</h2>
<p>Train and optimize our model using the training data with our custom feature set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lyseq_features_optimized</span> <span class="op">=</span> <span class="fu"><a href="../reference/DLBCLone_optimize_params.html">DLBCLone_optimize_params</a></span><span class="op">(</span>  </span>
<span>  <span class="va">mu_lyseq_train</span><span class="op">$</span><span class="va">features</span>, </span>
<span>  <span class="va">dlbcl_meta_clean</span>,</span>
<span>  umap_out <span class="op">=</span> <span class="va">mu_lyseq_train</span>,</span>
<span>  truth_classes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MCD"</span>,<span class="st">"EZB"</span>,<span class="st">"BN2"</span>,<span class="st">"N1"</span>,<span class="st">"ST2"</span>,<span class="st">"Other"</span><span class="op">)</span>,</span>
<span>  min_k<span class="op">=</span><span class="fl">11</span>,</span>
<span>  max_k<span class="op">=</span><span class="fl">21</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">lyseq_features_optimized</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DLBCLone_activate.html">DLBCLone_activate</a></span><span class="op">(</span> <span class="co"># Activates optimized model setting the UMAP projection</span></span>
<span>  <span class="va">lyseq_features_optimized</span>, </span>
<span>  force <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lyseq_features_optimized</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DLBCLone_activate.html">DLBCLone_activate</a></span><span class="op">(</span> <span class="co"># Activates optimized model setting the UMAP projection</span></span>
<span>  <span class="va">lyseq_features_optimized</span>, </span>
<span>  force <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>transforming each data point individually using the provided UMAP model. This will take some time.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code><span><span class="va">done</span></span></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>model given to function</code></pre>
</div>
</div>
</section><section class="level2"><h2 id="training-accuracy">Training accuracy<a class="anchor" aria-label="anchor" href="#training-accuracy"></a>
</h2>
<p>Check the classification accuracy of the optimized classifier (DLBCLone_wo) on the training data set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_alluvial.html">make_alluvial</a></span><span class="op">(</span><span class="va">lyseq_features_optimized</span>,</span>
<span>  pred_column<span class="op">=</span><span class="st">"DLBCLone_wo"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Duplicated aesthetics after name standardisation: fill</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in to_lodes_form(data = data, axes = axis_ind, discern =
params$discern): Some strata appear at multiple axes.
Warning in to_lodes_form(data = data, axes = axis_ind, discern =
params$discern): Some strata appear at multiple axes.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure><p><img src="validation_files/figure-html/alluvial%20train-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="sanity-check-feature-incidence-in-the-test-data">Sanity check feature incidence in the <em>test</em> data<a class="anchor" aria-label="anchor" href="#sanity-check-feature-incidence-in-the-test-data"></a>
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_check</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DLBCLone_predict.html">DLBCLone_predict</a></span><span class="op">(</span></span>
<span>  mutation_status <span class="op">=</span> <span class="va">lyseq_valid</span>,</span>
<span>  optimized_model <span class="op">=</span> <span class="va">lyseq_features_optimized</span>,</span>
<span>  check_frequencies <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_check</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="validation_files/figure-html/unnamed-chunk-3-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="apply-the-model-to-the-test-data">Apply the model to the <em>test</em> data<a class="anchor" aria-label="anchor" href="#apply-the-model-to-the-test-data"></a>
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_valid_lyseq_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DLBCLone_predict.html">DLBCLone_predict</a></span><span class="op">(</span></span>
<span>  mutation_status <span class="op">=</span> <span class="va">lyseq_valid</span>,</span>
<span>  optimized_model <span class="op">=</span> <span class="va">lyseq_features_optimized</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pred_valid_lyseq_all</span><span class="op">$</span><span class="va">prediction</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<table class="table caption-top" style="width:100%;">
<colgroup>
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 17%">
<col style="width: 5%">
<col style="width: 7%">
<col style="width: 5%">
<col style="width: 30%">
<col style="width: 0%">
<col style="width: 3%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 1%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
<col style="width: 0%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">predicted_label</th>
<th style="text-align: right;">confidence</th>
<th style="text-align: right;">other_score</th>
<th style="text-align: left;">neighbor_id</th>
<th style="text-align: left;">neighbor</th>
<th style="text-align: left;">distance</th>
<th style="text-align: left;">label</th>
<th style="text-align: left;">other_neighbor</th>
<th style="text-align: left;">vote_labels</th>
<th style="text-align: left;">weighted_votes</th>
<th style="text-align: right;">neighbors_other</th>
<th style="text-align: right;">neighborhood_otherness</th>
<th style="text-align: right;">other_weighted_votes</th>
<th style="text-align: right;">total_w</th>
<th style="text-align: right;">pred_w</th>
<th style="text-align: right;">V1</th>
<th style="text-align: right;">V2</th>
<th style="text-align: right;">.id</th>
<th style="text-align: right;">MCD_NN_count</th>
<th style="text-align: right;">EZB_NN_count</th>
<th style="text-align: right;">BN2_NN_count</th>
<th style="text-align: right;">N1_NN_count</th>
<th style="text-align: right;">ST2_NN_count</th>
<th style="text-align: right;">Other_NN_count</th>
<th style="text-align: left;">top_group</th>
<th style="text-align: right;">MCD_score</th>
<th style="text-align: right;">EZB_score</th>
<th style="text-align: right;">BN2_score</th>
<th style="text-align: right;">N1_score</th>
<th style="text-align: right;">ST2_score</th>
<th style="text-align: right;">Other_score</th>
<th style="text-align: left;">top_score_group</th>
<th style="text-align: right;">top_group_score</th>
<th style="text-align: right;">top_group_count</th>
<th style="text-align: right;">Other_count</th>
<th style="text-align: right;">by_vote</th>
<th style="text-align: left;">by_vote_opt</th>
<th style="text-align: left;">by_score</th>
<th style="text-align: right;">score_ratio</th>
<th style="text-align: left;">by_score_opt</th>
<th style="text-align: left;">DLBCLone_w</th>
<th style="text-align: left;">DLBCLone_wo</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">00-14595_tumorA</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">05-25439T,17-33848_tumorB,DLBCL-MC_F349_BJJ-Tumor,06-18449_tumorB,05-25674T,CAR_276_PreCART,DLBCL11572T,SP194216,SP193546,DLBCL-MC_F358_WDS-Tumor,DLBCL11666T,02-13135T,CAR_154_PreCART,00-14595_tumorC,Aph_13_PreCART,99-13280T,DLBCL10465T,DLBCL10464T,03-11110T</td>
<td style="text-align: left;">88,354,626,109,89,526,1218,1301,1297,627,1249,29,508,1,469,447,837,836,47</td>
<td style="text-align: left;">0.129,0.147,0.171,0.173,0.176,0.181,0.19,0.201,0.202,0.212,0.213,0.22,0.237,0.255,0.259,0.292,0.306,0.32,0.325</td>
<td style="text-align: left;">EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">91.4646166144537</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: right;">91.46462</td>
<td style="text-align: right;">91.46462</td>
<td style="text-align: right;">1.964856</td>
<td style="text-align: right;">4.0599666</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">91.464617</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">91.46462</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">19</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">Inf</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
</tr>
<tr class="even">
<td style="text-align: left;">00-14595_tumorB</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">17-33848_tumorB,02-13135T,CAR_154_PreCART,SP193546,DLBCL-MC_F349_BJJ-Tumor,05-25439T,CAR_276_PreCART,00-14595_tumorC,06-18449_tumorB,05-25674T,DLBCL-MC_F358_WDS-Tumor,DLBCL11572T,SP194216,DLBCL11666T,13-26601T,Aph_13_PreCART,SP116659,06-30145T,SP192993</td>
<td style="text-align: left;">354,29,508,1297,626,88,526,1,109,89,627,1218,1301,1249,246,469,1339,122,1281</td>
<td style="text-align: left;">0.141,0.154,0.169,0.174,0.183,0.189,0.199,0.207,0.209,0.246,0.251,0.261,0.267,0.285,0.305,0.328,0.328,0.338,0.346</td>
<td style="text-align: left;">EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">84.9552949261059</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: right;">84.95529</td>
<td style="text-align: right;">84.95529</td>
<td style="text-align: right;">2.035857</td>
<td style="text-align: right;">4.0529613</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">84.955295</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">84.95529</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">19</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">Inf</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
</tr>
<tr class="odd">
<td style="text-align: left;">00-14595_tumorC</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">SP193546,02-13135T,13-26601T,SP116659,CAR_154_PreCART,SP192993,08-13706T,06-18449_tumorB,17-33848_tumorB,05-25439T,CLC03456,CAR_178_PreCART,DLBCL-MC_F300_TBF-Tumor,05-24666T,SU-DHL-6,DLBCL-MC_F349_BJJ-Tumor,DLBCL11532T,CAR_276_PreCART,05-25674T</td>
<td style="text-align: left;">1296,28,245,1338,507,1280,139,108,353,87,368,514,621,85,2424,625,1194,525,88</td>
<td style="text-align: left;">0.177,0.179,0.196,0.214,0.239,0.244,0.249,0.255,0.267,0.275,0.295,0.296,0.301,0.308,0.309,0.31,0.322,0.327,0.33</td>
<td style="text-align: left;">EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">73.5868644142581</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: right;">73.58686</td>
<td style="text-align: right;">73.58686</td>
<td style="text-align: right;">2.094698</td>
<td style="text-align: right;">3.9400454</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">73.586864</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">73.58686</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">19</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">Inf</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
</tr>
<tr class="even">
<td style="text-align: left;">00-14595_tumorD</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">SP194216,DLBCL-MC_F358_WDS-Tumor,DLBCL11572T,Aph_13_PreCART,DLBCL11666T,CAR_276_PreCART,DLBCL-MC_F349_BJJ-Tumor,05-25674T,DLBCL11573T,SP116620,05-25439T,06-30025T,17-33848_tumorB,190806-PL01-PL02,99-13280T,06-18449_tumorB,16-13732T,DLBCL-DFCI_DLBCL_Goe16-Tumor,06-30145T</td>
<td style="text-align: left;">1301,627,1218,469,1249,526,626,89,1219,1331,88,121,354,450,447,109,321,612,122</td>
<td style="text-align: left;">0.137,0.15,0.178,0.178,0.179,0.193,0.205,0.207,0.221,0.223,0.229,0.229,0.234,0.26,0.262,0.273,0.286,0.301,0.301</td>
<td style="text-align: left;">EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB,EZB</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">89.1139107432683</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: right;">89.11391</td>
<td style="text-align: right;">89.11391</td>
<td style="text-align: right;">1.892516</td>
<td style="text-align: right;">4.1527591</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">89.113911</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">89.11391</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">19</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: right;">Inf</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
<td style="text-align: left;">EZB</td>
</tr>
<tr class="odd">
<td style="text-align: left;">00-15201_tumorA</td>
<td style="text-align: left;">BN2</td>
<td style="text-align: right;">0.7997312</td>
<td style="text-align: right;">2.2781338</td>
<td style="text-align: left;">190017-PL01,Reddy_3551T,Reddy_3482T,DLBCL-c_D_1117-Tumor,DLBCL-c_D_1158-Tumor,SP194228,Reddy_2166T,DLBCL10969T,DLBCL10953T,Reddy_2716T,Reddy_689T,EX01025,Reddy_3515T,Reddy_2214T,DLBCL10927T,DLBCL-RICOVER_711-Tumor,Reddy_3873T,DLBCL11677T,EX01023</td>
<td style="text-align: left;">464,2075,2028,777,813,1301,1599,1045,1029,1831,2404,586,2052,1620,1007,741,2288,1257,584</td>
<td style="text-align: left;">0.196,0.229,0.38,0.396,0.445,0.496,0.501,0.51,0.515,0.52,0.525,0.546,0.627,0.691,0.697,0.702,0.705,0.817,0.819</td>
<td style="text-align: left;">BN2,BN2,BN2,ST2,ST2,BN2,BN2,BN2,BN2,BN2,BN2,EZB,BN2,EZB,BN2,BN2,BN2,BN2,BN2</td>
<td style="text-align: left;">Reddy_2246T,Reddy_3845T,DLBCL10499T,QC2-26T,1102503011T,CAR_109_PreCART,Reddy_3733T,03-10440_tumorB,03-10440_tumorA,Reddy_3461T,Reddy_2634T,DLBCL11486T,Reddy_3503T,Reddy_3843T,Reddy_3566T,Reddy_2919T,DLBCL11504T,Reddy_2206T,DLBCL11002T,DLBCL11681T,Reddy_3612T,SP59400,141900-PL01,1011-2018_FFPE,Reddy_2867T,1011-2018_plasma,Reddy_3418T,1001-2030_FFPE,DLBCL10955T,DLBCL10888T,DLBCL11595T,DLBCL-c_D_1162-Tumor,DLBCL10884T,DLBCL10485T,Reddy_790T,DLBCL11271T</td>
<td style="text-align: left;">BN2,EZB,ST2</td>
<td style="text-align: left;">32.1633840869331,3.27813883002644,4.77621933235038</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">1.8947368</td>
<td style="text-align: right;">73.272493</td>
<td style="text-align: right;">40.21774</td>
<td style="text-align: right;">32.16338</td>
<td style="text-align: right;">1.174061</td>
<td style="text-align: right;">-0.7965413</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">BN2</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3.278139</td>
<td style="text-align: right;">32.16338</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">4.776219</td>
<td style="text-align: right;">73.272493</td>
<td style="text-align: left;">BN2</td>
<td style="text-align: right;">32.16338</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">15</td>
<td style="text-align: left;">Other</td>
<td style="text-align: left;">BN2</td>
<td style="text-align: right;">0.4389558</td>
<td style="text-align: left;">Other</td>
<td style="text-align: left;">BN2</td>
<td style="text-align: left;">BN2</td>
</tr>
<tr class="even">
<td style="text-align: left;">00-15201_tumorB</td>
<td style="text-align: left;">N1</td>
<td style="text-align: right;">0.9052115</td>
<td style="text-align: right;">0.1549505</td>
<td style="text-align: left;">10-39294_tumorB,DLBCL10520T,DLBCL11648T,10-39294_tumorA,DLBCL-MAYO_DLBCL_234-Tumor,DLBCL11005T,DLBCL11439T,DLBCL11513T,14-33798_tumorB,c_D_1050_T,DLBCL11503T,DLBCL10973T,DLBCL11564T,DLBCL11203T,DLBCL10844T,1006-2010_FFPE,Reddy_3997T,04-35039T,DLBCL10952T</td>
<td style="text-align: left;">187,885,1235,186,1442,1078,1115,1180,276,1433,1172,1048,1209,1088,929,210,2397,70,1028</td>
<td style="text-align: left;">0.274,0.415,0.447,0.469,0.493,0.56,0.573,0.602,0.659,0.691,0.71,0.719,0.754,0.763,0.772,0.777,0.787,0.795,0.796</td>
<td style="text-align: left;">N1,N1,N1,N1,N1,N1,ST2,N1,N1,N1,N1,N1,ST2,N1,N1,N1,N1,N1,N1</td>
<td style="text-align: left;">1102503029T,Reddy_3448T,DLBCL10837T</td>
<td style="text-align: left;">N1,ST2</td>
<td style="text-align: left;">29.3438944567235,3.07272235128703</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0.1578947</td>
<td style="text-align: right;">4.546851</td>
<td style="text-align: right;">32.41662</td>
<td style="text-align: right;">29.34389</td>
<td style="text-align: right;">-1.056905</td>
<td style="text-align: right;">-3.6693215</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">N1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.000000</td>
<td style="text-align: right;">0.00000</td>
<td style="text-align: right;">29.34389</td>
<td style="text-align: right;">3.072722</td>
<td style="text-align: right;">4.546851</td>
<td style="text-align: left;">N1</td>
<td style="text-align: right;">29.34389</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">17</td>
<td style="text-align: left;">N1</td>
<td style="text-align: left;">N1</td>
<td style="text-align: right;">6.4536741</td>
<td style="text-align: left;">N1</td>
<td style="text-align: left;">N1</td>
<td style="text-align: left;">N1</td>
</tr>
</tbody>
</table>
</div>
</div>
<section class="level3"><h3 id="sanity-check-the-umap">Sanity check the UMAP<a class="anchor" aria-label="anchor" href="#sanity-check-the-umap"></a>
</h3>
<p>If the UMAP is working as it should be, you should see a similar level of separation of your <em>test</em> samples in UMAP space. Since we have LymphGen classifications for our test samples, we can check this. In order to visualize these, the metadata will need to be attached to the UMAP projection of the new samples. What you should hopefully see is roughly the same placement of your samples on this scatterplot relative to the training samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">projected_with_meta</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">pred_valid_lyseq_all</span><span class="op">$</span><span class="va">projection</span>,</span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">valid_metadata</span>,<span class="va">sample_id</span>,<span class="va">lymphgen</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(sample_id)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">projected_with_meta</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        sample_id        V1         V2 lymphgen
1 00-14595_tumorA  1.964856  4.0599666      EZB
2 00-14595_tumorB  2.035857  4.0529613      EZB
3 00-14595_tumorC  2.094698  3.9400454      EZB
4 00-14595_tumorD  1.892516  4.1527591      EZB
5 00-15201_tumorA  1.174061 -0.7965413    Other
6 00-15201_tumorB -1.056905 -3.6693215       N1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">projected_with_meta</span><span class="op">$</span><span class="va">lymphgen</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "EZB"         "Other"       "N1"          "BN2/N1"      "BN2"
 [6] "MCD"         "ST2"         "EZB/MCD/ST2" "EZB/ST2"     "BN2/EZB"
[11] "BN2/MCD"     "EZB/MCD"    </code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#drop composites</span></span>
<span><span class="va">projected_with_meta</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">projected_with_meta</span>,<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"/"</span>,<span class="va">lymphgen</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_umap_scatterplot.html">make_umap_scatterplot</a></span><span class="op">(</span><span class="va">projected_with_meta</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="validation_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level3"><h3 id="make_neighborhood_plot">make_neighborhood_plot<a class="anchor" aria-label="anchor" href="#make_neighborhood_plot"></a>
</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_sample</span> <span class="op">=</span> <span class="st">"92-38267_tumorB"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/make_neighborhood_plot.html">make_neighborhood_plot</a></span><span class="op">(</span></span>
<span>  single_sample_prediction_output <span class="op">=</span> <span class="va">pred_valid_lyseq_all</span>,</span>
<span>  this_sample_id <span class="op">=</span> <span class="va">test_sample</span>,</span>
<span>  prediction_in_title <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  add_circle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label_column <span class="op">=</span> <span class="st">"DLBCLone_wo"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="validation_files/figure-html/neighborhood-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level3"><h3 id="nearest_neighbor_heatmap">nearest_neighbor_heatmap<a class="anchor" aria-label="anchor" href="#nearest_neighbor_heatmap"></a>
</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nearest_neighbor_heatmap.html">nearest_neighbor_heatmap</a></span><span class="op">(</span></span>
<span>  this_sample_id <span class="op">=</span> <span class="va">test_sample</span>,</span>
<span>  DLBCLone_model <span class="op">=</span> <span class="va">pred_valid_lyseq_all</span>,</span>
<span>  font_size <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="validation_files/figure-html/heatmap-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section class="level2"><h2 id="accuracy-assessment">Accuracy Assessment<a class="anchor" aria-label="anchor" href="#accuracy-assessment"></a>
</h2>
<p>Finally, lets assess the accuracy of the validation set by comparing <code>dlbclone_predict</code> predictions with the LymphGen results that were previously generated.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_vs_original</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">pred_valid_lyseq_all</span><span class="op">$</span><span class="va">prediction</span>,<span class="va">valid_metadata</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#collapse composites into group-COMP</span></span>
<span><span class="va">pred_vs_original</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.utils/man/tidy_lymphgen.html" class="external-link">tidy_lymphgen</a></span><span class="op">(</span><span class="va">pred_vs_original</span>,lymphgen_column_in <span class="op">=</span> <span class="st">"lymphgen"</span>,lymphgen_column_out <span class="op">=</span> <span class="st">"lymphgen_tidy"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#convert composites into their primary class</span></span>
<span><span class="va">pred_vs_original</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">pred_vs_original</span>,lymphgen_basic <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"-COMP"</span>,<span class="st">""</span>,<span class="va">lymphgen_tidy</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/make_alluvial.html">make_alluvial</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>predictions<span class="op">=</span><span class="va">pred_vs_original</span><span class="op">)</span>,</span>
<span>  pred_column <span class="op">=</span> <span class="st">"DLBCLone_wo"</span>,</span>
<span>  group_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MCD"</span>,<span class="st">"BN2"</span>,<span class="st">"ST2"</span>,<span class="st">"EZB"</span>,<span class="st">"N1"</span>,<span class="st">"Other"</span><span class="op">)</span>,</span>
<span>  truth_column <span class="op">=</span> <span class="st">"lymphgen_basic"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="validation_files/figure-html/compare_to_lymphgen-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<section class="level3"><h3 id="oncoplot">Oncoplot<a class="anchor" aria-label="anchor" href="#oncoplot"></a>
</h3>
<p>Another way to visualize the results is via an Oncoplot (using GAMBLR.viz::prettyOncoplot).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://morinlab.github.io/GAMBLR.viz/" class="external-link">GAMBLR.viz</a></span><span class="op">)</span></span>
<span><span class="va">ashm_genes</span> <span class="op">=</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">ashm_genes</span>,<span class="va">lyseq_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="va">silent_show</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Silent"</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ashm_genes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">silent_show</span><span class="op">)</span><span class="op">=</span><span class="va">ashm_genes</span></span>
<span><span class="co"># add hotspot annotation</span></span>
<span><span class="va">valid_maf_vaf</span><span class="op">$</span><span class="va">hot_spot</span> <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="va">valid_maf_vaf</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">valid_maf_vaf</span>,hot_spot<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span><span class="op">==</span><span class="st">"MYD88"</span> <span class="op">&amp;</span> <span class="va">HGVSp_Short</span><span class="op">==</span><span class="st">"p.L265P"</span>,<span class="cn">TRUE</span>,<span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://morinlab.github.io/GAMBLR.viz/reference/prettyOncoplot.html" class="external-link">prettyOncoplot</a></span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">pred_vs_original</span>, </span>
<span>  maf_df <span class="op">=</span> <span class="va">valid_maf_vaf</span>,</span>
<span>  metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lymphgen_basic"</span>,<span class="st">"DLBCLone_w"</span>,<span class="st">"DLBCLone_wo"</span><span class="op">)</span>,</span>
<span>  minMutationPercent <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  highlightHotspots <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  genes <span class="op">=</span> <span class="va">lyseq_genes</span>,</span>
<span>  pctFontSize <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  fontSizeGene <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  hideSideBarplot <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  simplify_annotation <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  hide_annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCLone_wo"</span>,<span class="st">"DLBCLone_w"</span><span class="op">)</span>,</span>
<span>  include_noncoding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">silent_show</span><span class="op">)</span>,</span>
<span>  sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCLone_wo"</span>,<span class="st">"lymphgen_basic"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="validation_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kostiantyn Dreval, Ryan Morin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
