<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Assemble genetic features for UMAP input — assemble_genetic_features • GAMBLR.predict</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Assemble genetic features for UMAP input — assemble_genetic_features"><meta name="description" content="This function assembles a matrix of genetic features for each sample,
including mutation status, aSHM counts, and structural variant status
for BCL2, BCL6, and MYC. It supports both genome and capture
sequencing (data) types."><meta property="og:description" content="This function assembles a matrix of genetic features for each sample,
including mutation status, aSHM counts, and structural variant status
for BCL2, BCL6, and MYC. It supports both genome and capture
sequencing (data) types."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.predict</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/index.html">DLBCLone: Overview</a></li>
    <li><a class="dropdown-item" href="../articles/umap.html">Exploring and selecting features</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_optimize_params.html">Training a new DLBCLone model</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_stored.html">DLBCLone Models</a></li>
    <li><a class="dropdown-item" href="../articles/predict_samples.html">Classifying samples with DLBCLone</a></li>
    <li><a class="dropdown-item" href="../articles/full_walkthrough.html">DLBCLone: Full Walkthrough</a></li>
    <li><a class="dropdown-item" href="../articles/validation.html">Applying a DLBCLone model to new data</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_KNN.html">K-Nearest Neighbors without UMAP</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_mixture_model.html">Gaussian Mixture Model</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Assemble genetic features for UMAP input</h1>

      <div class="d-none name"><code>assemble_genetic_features.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function assembles a matrix of genetic features for each sample,
including mutation status, aSHM counts, and structural variant status
for BCL2, BCL6, and MYC. It supports both genome and capture
sequencing (data) types.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">assemble_genetic_features</span><span class="op">(</span></span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  sv_from_metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>BCL2 <span class="op">=</span> <span class="st">"bcl2_ba"</span>, BCL6 <span class="op">=</span> <span class="st">"bcl6_ba"</span>, MYC <span class="op">=</span> <span class="st">"myc_ba"</span><span class="op">)</span>,</span>
<span>  <span class="va">genes</span>,</span>
<span>  <span class="va">synon_genes</span>,</span>
<span>  <span class="va">maf_with_synon</span>,</span>
<span>  <span class="va">hotspot_genes</span>,</span>
<span>  genome_build <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  sv_value <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  synon_value <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  coding_value <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  include_ashm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">annotated_sv</span>,</span>
<span>  include_GAMBL_sv <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  review_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>Data frame with sample metadata,
must include seq_type and sample_id.</p></dd>


<dt id="arg-sv-from-metadata">sv_from_metadata<a class="anchor" aria-label="anchor" href="#arg-sv-from-metadata"></a></dt>
<dd><p>A named vector that specifies the columns
containing the oncogene
translocation status for any SV that is annotated in the metadata
Where the name is the oncogene and the value is the column name in
the metadata.
The column created in the output will be "<oncogenename>_SV".</oncogenename></p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>Vector of gene symbols to include.</p></dd>


<dt id="arg-synon-genes">synon_genes<a class="anchor" aria-label="anchor" href="#arg-synon-genes"></a></dt>
<dd><p>Vector of gene symbols for synonymous mutations
(generally a subset of genes).</p></dd>


<dt id="arg-maf-with-synon">maf_with_synon<a class="anchor" aria-label="anchor" href="#arg-maf-with-synon"></a></dt>
<dd><p>MAF data frame including synonymous mutations.</p></dd>


<dt id="arg-hotspot-genes">hotspot_genes<a class="anchor" aria-label="anchor" href="#arg-hotspot-genes"></a></dt>
<dd><p>Vector specifying genes for which hotspot
mutations should be separately annotated.
The columns will be named "<gene>HOTSPOT". For this to work,
either specify review_hotspots = TRUE or, if you want full control over
hotspot annotation, the MAF must include a column "hot_spot" with TRUE
specifying any row corresponding to a hotspot mutation.</gene></p></dd>


<dt id="arg-sv-value">sv_value<a class="anchor" aria-label="anchor" href="#arg-sv-value"></a></dt>
<dd><p>Value to assign for SV presence (default: 2).</p></dd>


<dt id="arg-synon-value">synon_value<a class="anchor" aria-label="anchor" href="#arg-synon-value"></a></dt>
<dd><p>Value to assign for synonymous mutations (default: 1).</p></dd>


<dt id="arg-coding-value">coding_value<a class="anchor" aria-label="anchor" href="#arg-coding-value"></a></dt>
<dd><p>Value to assign for coding mutations (default: 2).</p></dd>


<dt id="arg-include-ashm">include_ashm<a class="anchor" aria-label="anchor" href="#arg-include-ashm"></a></dt>
<dd><p>Logical; if TRUE, use GAMBLR.results::get_ssm_by_region
to retrieve all non-coding mutations for each gene in synon_genes and
use these to infer mutation status (default: FALSE).
WARNING: This feature is experimental and is likely not going to give
comparable results if you are using a mix of genome and capture data.
It also relies on GAMBLR.results, which is not a core dependency of
GAMBLR.predict.</p></dd>


<dt id="arg-annotated-sv">annotated_sv<a class="anchor" aria-label="anchor" href="#arg-annotated-sv"></a></dt>
<dd><p>Data frame in bedpe format with annotated SVs from
GAMBLR.utils::annotate_sv(). If provided, the oncogene SV status will
be based on the union of the SVs in this data frame and the metadata
columns specified in sv_from_metadata.</p></dd>


<dt id="arg-include-gambl-sv">include_GAMBL_sv<a class="anchor" aria-label="anchor" href="#arg-include-gambl-sv"></a></dt>
<dd><p>Logical; if TRUE, SVs from GAMBLR.results will
automatically be retrieved and annotated.
WARNING: This feature is experimental and is likely not going to give
comparable results if you are using a mix of genome and capture data.
It also relies on GAMBLR.results, which is not a core dependency of
GAMBLR.predict.</p></dd>


<dt id="arg-review-hotspots">review_hotspots<a class="anchor" aria-label="anchor" href="#arg-review-hotspots"></a></dt>
<dd><p>Logical; if TRUE, any gene in hotspot_genes
that is compatible with review_hotspots will have its hotspots
annotated. For more information, see GAMBLR.helpers::review_hotspots</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Defaults to FALSE</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Matrix of assembled features for each sample.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">all_meta</span> <span class="op">=</span> <span class="fu">get_gambl_metadata</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span>,<span class="va">seq_type</span><span class="op">==</span><span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">all_maf</span> <span class="op">=</span> <span class="fu">get_all_coding_ssm</span><span class="op">(</span><span class="va">all_meta</span>,include_silent<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sv_all</span> <span class="op">=</span><span class="fu">get_combined_sv</span><span class="op">(</span><span class="va">all_meta</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">anno_sv</span> <span class="op">=</span> <span class="fu">annotate_sv</span><span class="op">(</span><span class="va">sv_all</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">feat_mat</span> <span class="op">=</span> <span class="fu">assemble_genetic_features</span><span class="op">(</span><span class="va">all_meta</span>,</span></span>
<span class="r-in"><span>                                    genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EZH2"</span>,<span class="st">"SOCS1"</span>,<span class="st">"PIM1"</span>,</span></span>
<span class="r-in"><span>                                            <span class="st">"MYD88"</span>,<span class="st">"CREBBP"</span>,<span class="st">"SGK1"</span>,</span></span>
<span class="r-in"><span>                                            <span class="st">"NOTCH2"</span>,<span class="st">"NOTCH1"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                    maf_with_synon <span class="op">=</span> <span class="va">all_maf</span>,</span></span>
<span class="r-in"><span>                                    annotated_sv <span class="op">=</span> <span class="va">anno_sv</span>,</span></span>
<span class="r-in"><span>                                    synon_genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PIM1"</span>,<span class="st">"SOCS1"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kostiantyn Dreval, Ryan Morin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

