<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Assign CN to SSM. — assign_cn_to_ssm • GAMBLR.open</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Assign CN to SSM. — assign_cn_to_ssm"><meta name="description" content="Annotate mutations with their copy number information."><meta property="og:description" content="Annotate mutations with their copy number information."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.open</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-open.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/mutation_summary.html">Exploring simple somatic mutations</a></li>
    <li><a class="dropdown-item" href="../articles/feature_matrix.html">Summarizing mutation status as a feature matrix</a></li>
    <li><a class="dropdown-item" href="../articles/genetic_classification.html">Genetic classification of tumours</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Assign CN to SSM.</h1>

      <div class="d-none name"><code>assign_cn_to_ssm.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Annotate mutations with their copy number information.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">assign_cn_to_ssm</span><span class="op">(</span></span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  <span class="va">seg_data</span>,</span>
<span>  <span class="va">projection</span>,</span>
<span>  coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  assume_diploid <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>Metadata table with one or more rows to specify
the samples to process.</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>A data frame of mutations in MAF format or maf_data object
(e.g. from <code>get_coding_ssm</code> or <code>get_ssm_by_sample</code>).</p></dd>


<dt id="arg-seg-data">seg_data<a class="anchor" aria-label="anchor" href="#arg-seg-data"></a></dt>
<dd><p>A data frame of segmented copy number data or seg_data object</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>Specified genome projection that returned data is relative to.
This is only required when it cannot be inferred from maf_df or seg_df
(or they are not provided).</p></dd>


<dt id="arg-coding-only">coding_only<a class="anchor" aria-label="anchor" href="#arg-coding-only"></a></dt>
<dd><p>Optional. Set to TRUE to restrict to only variants in coding space
Default is to work with genome-wide variants.</p></dd>


<dt id="arg-assume-diploid">assume_diploid<a class="anchor" aria-label="anchor" href="#arg-assume-diploid"></a></dt>
<dd><p>Optional, this parameter annotates every mutation as
copy neutral. Default is FALSE.</p></dd>


<dt id="arg-include-silent">include_silent<a class="anchor" aria-label="anchor" href="#arg-include-silent"></a></dt>
<dd><p>Logical parameter indicating whether to include silent
mutations in coding space. Default is FALSE. This parameter only
makes sense if <code>coding_only</code> is set to TRUE.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Any additional parameters.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing a data frame (MAF-like format) with three extra
columns:
- log.ratio is the log ratio from the seg file (NA when no overlap).
- LOH
- CN (the rounded absolute copy number estimate of the region based on
log.ratio, NA when no overlap was found).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function takes a metadata table and returns all mutations
for the samples in that metadata. Each mutation is annotated with the
local copy number state of each mutated site. The user can specify if
only coding mutations are of interest. To do so,
set <code>coding_only = TRUE</code>. When necessary, this function relies on
<code>get_ssm_by_samples</code> and <code>get_cn_segments</code> to obtain the required data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># long-handed way</span></span></span>
<span class="r-in"><span><span class="co"># 1. get some metadata for a collection of samples</span></span></span>
<span class="r-in"><span><span class="va">some_meta</span> <span class="op">=</span> <span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">study</span><span class="op">==</span><span class="st">"FL_Dreval"</span>,</span></span>
<span class="r-in"><span>        <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"SP"</span>,<span class="va">sample_id</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled metadata in GAMBLR.data...</span>
<span class="r-in"><span><span class="co"># 2. Get the SSMs for these samples</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ssm_genomes_grch37</span> <span class="op">=</span> <span class="fu"><a href="get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>projection <span class="op">=</span> <span class="st">"grch37"</span>,</span></span>
<span class="r-in"><span>                                  these_samples_metadata <span class="op">=</span> <span class="va">some_meta</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled SSM calls (.maf) calls in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> after linking with metadata, we have mutations from 182 samples</span>
<span class="r-in"><span><span class="co"># peek at the results</span></span></span>
<span class="r-in"><span><span class="va">ssm_genomes_grch37</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       HNRNPU              0      .     GRCh37          1      245027102</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2         PCLO              0      .     GRCh37          7       82784068</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        MYOM2              0      .     GRCh37          8        2063762</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       CREBBP              0      .     GRCh37         16        3786740</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        IGLL5              0      .     GRCh37         22       23230399</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    HIST1H2BC              0      .     GRCh37          6       26123774</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7        SETD2              0      .     GRCh37          3       47088059</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8         IRF8              0      .     GRCh37         16       85942737</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9       ARID5B              0      .     GRCh37         10       63661999</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10      ARID5B              0      .     GRCh37         10       63810677</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    End_Position Strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     245027102      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      82784068      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       2063762      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       3786740      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      23230400      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      26123774      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7      47088059      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8      85942737      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9      63661999      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10     63810677      +</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 3. Lazily let this function obtain the corresponding seg_data</span></span></span>
<span class="r-in"><span><span class="co"># for the right genome_build</span></span></span>
<span class="r-in"><span><span class="va">cn_list</span> <span class="op">=</span> <span class="fu">assign_cn_to_ssm</span><span class="op">(</span><span class="va">some_meta</span>,<span class="va">ssm_genomes_grch37</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled CN segments (.seg) calls in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running in default mode of any...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cn_list</span><span class="op">$</span><span class="va">maf</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>,<span class="va">log.ratio</span>,<span class="va">CN</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       HNRNPU              0      .     GRCh37          1      245027102</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2         PCLO              0      .     GRCh37          7       82784068</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        MYOM2              0      .     GRCh37          8        2063762</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       CREBBP              0      .     GRCh37         16        3786740</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        IGLL5              0      .     GRCh37         22       23230399</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    HIST1H2BC              0      .     GRCh37          6       26123774</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7        SETD2              0      .     GRCh37          3       47088059</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8         IRF8              0      .     GRCh37         16       85942737</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9       ARID5B              0      .     GRCh37         10       63661999</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10      ARID5B              0      .     GRCh37         10       63810677</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    End_Position Strand log.ratio CN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     245027102      + 0.4306932  3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      82784068      + 0.0000000  2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       2063762      + 0.0000000  2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       3786740      + 0.0000000  2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      23230400      + 0.0000000  2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      26123774      + 0.0000000  2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7      47088059      + 0.0000000  2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8      85942737      + 1.3219281  5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9      63661999      + 0.5849625  3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10     63810677      + 0.5849625  3</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># This wouldn't work because the hg38 seg_data is not bundled</span></span></span>
<span class="r-in"><span><span class="va">ssm_genomes_hg38</span> <span class="op">=</span> <span class="fu"><a href="get_coding_ssm.html">get_coding_ssm</a></span><span class="op">(</span>projection <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                                  these_samples_metadata <span class="op">=</span> <span class="va">some_meta</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">cn_list</span> <span class="op">=</span> <span class="fu">assign_cn_to_ssm</span><span class="op">(</span><span class="va">some_meta</span>,<span class="va">ssm_genomes_hg38</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Easiest/laziest way:</span></span></span>
<span class="r-in"><span><span class="va">cn_list</span> <span class="op">=</span> <span class="fu">assign_cn_to_ssm</span><span class="op">(</span>projection <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cn_list</span><span class="op">$</span><span class="va">maf</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Tumor_Sample_Barcode</span>,<span class="va">CN</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostia Dreval.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

