<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get CNV and coding SSM combined status — get_cnv_and_ssm_status • GAMBLR.open</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get CNV and coding SSM combined status — get_cnv_and_ssm_status"><meta name="description" content="For each specified chromosome region (gene name), return status 1 if the copy number (CN)
state is non-neutral, i.e. different from 2, or if the region contains any coding simple somatic mutation (SSM)."><meta property="og:description" content="For each specified chromosome region (gene name), return status 1 if the copy number (CN)
state is non-neutral, i.e. different from 2, or if the region contains any coding simple somatic mutation (SSM)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.open</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-open.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/mutation_summary.html">Exploring simple somatic mutations</a></li>
    <li><a class="dropdown-item" href="../articles/feature_matrix.html">Summarizing mutation status as a feature matrix</a></li>
    <li><a class="dropdown-item" href="../articles/genetic_classification.html">Genetic classification of tumours</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get CNV and coding SSM combined status</h1>

      <div class="d-none name"><code>get_cnv_and_ssm_status.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>For each specified chromosome region (gene name), return status 1 if the copy number (CN)
state is non-neutral, <em>i.e.</em> different from 2, or if the region contains any coding simple somatic mutation (SSM).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_cnv_and_ssm_status</span><span class="op">(</span></span>
<span>  <span class="va">genes_and_cn_threshs</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">maf_df</span>,</span>
<span>  <span class="va">seg_data</span>,</span>
<span>  <span class="va">cn_matrix</span>,</span>
<span>  only_cnv <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  genome_build <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  include_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  review_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  adjust_for_ploidy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  include_silent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">this_seq_type</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-genes-and-cn-threshs">genes_and_cn_threshs<a class="anchor" aria-label="anchor" href="#arg-genes-and-cn-threshs"></a></dt>
<dd><p>A data frame with columns "gene_id" and "cn_thresh". The "gene_id" column stores
gene symbols (characters) which determine the regions to return CNV and/or coding SSM status. The "cn_thresh"
column stores integers that mean the maximum or minimum CN states to return status 1 (contains CNV) for
its respective gene. If this integer is below 2 (neutral CN state for diploids), it is taken as the maximum
(gene consider as tumor suppressor); if above 2, it is the minimum (oncogene); if equal to 2, do not consider
CNV to return status.</p></dd>


<dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>The metadata for samples of interest to be included in the returned matrix.
Can be created with <code>get_gambl_metadata</code> function.</p></dd>


<dt id="arg-maf-df">maf_df<a class="anchor" aria-label="anchor" href="#arg-maf-df"></a></dt>
<dd><p>Optional data frame containing the coding variants for your samples (i.e. output from <code>get_all_coding_ssm</code>)</p></dd>


<dt id="arg-seg-data">seg_data<a class="anchor" aria-label="anchor" href="#arg-seg-data"></a></dt>
<dd><p>Optionally provide the function with a data frame of segments that will be used instead of the GAMBL flatfiles</p></dd>


<dt id="arg-cn-matrix">cn_matrix<a class="anchor" aria-label="anchor" href="#arg-cn-matrix"></a></dt>
<dd><p>Instead of seg_data, you can provide a matrix of CN values for the samples in the metadata.
See <a href="https://rdrr.io/pkg/GAMBLR.utils/man/segmented_data_to_cn_matrix.html" class="external-link">GAMBLR.utils::segmented_data_to_cn_matrix</a> for more information on how to create this matrix.</p></dd>


<dt id="arg-only-cnv">only_cnv<a class="anchor" aria-label="anchor" href="#arg-only-cnv"></a></dt>
<dd><p>A vector of gene names indicating the genes for which only CNV status should be considered,
ignoring SSM status. Set this argument to "all" or "none" (default) to apply this behavior to all or none
of the genes, respectively.</p></dd>


<dt id="arg-genome-build">genome_build<a class="anchor" aria-label="anchor" href="#arg-genome-build"></a></dt>
<dd><p>Reference genome build. Possible values are "grch37" (default) or "hg38".</p></dd>


<dt id="arg-include-hotspots">include_hotspots<a class="anchor" aria-label="anchor" href="#arg-include-hotspots"></a></dt>
<dd><p>Logical parameter indicating whether hotspots object should also be tabulated. Default is TRUE.</p></dd>


<dt id="arg-review-hotspots">review_hotspots<a class="anchor" aria-label="anchor" href="#arg-review-hotspots"></a></dt>
<dd><p>Logical parameter indicating whether hotspots object should be reviewed to include
functionally relevant mutations or rare lymphoma-related genes. Default is TRUE.</p></dd>


<dt id="arg-adjust-for-ploidy">adjust_for_ploidy<a class="anchor" aria-label="anchor" href="#arg-adjust-for-ploidy"></a></dt>
<dd><p>Set to FALSE to disable scaling of CN values by the genome-wide average per sample</p></dd>


<dt id="arg-include-silent">include_silent<a class="anchor" aria-label="anchor" href="#arg-include-silent"></a></dt>
<dd><p>Set to TRUE if you want Synonymous mutations to also be considered</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>Deprecated</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with CNV and SSM combined status.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The user can choose from which regions are intended to return only copy number variation (CNV) status,
only coding SSM status, or at least the presence of one of them. This behavior is controlled by the arguments
<code>genes_and_cn_threshs</code> (column <code>cn_thresh</code>) and <code>only_cnv</code>.</p>
<p>This function internally calls the <code>get_cn_states</code>, <code>get_ssm_by_samples</code> and <code>get_coding_ssm_status</code>functions.
Therefore, many of its arguments are assigned to these functions. If needed, see the documentation of these
functions for more information.</p>
<p>In the case of returning NA values, this is due to the <code>get_cn_segments</code> function not being able to internally
return any copy number segments from the specified chromosome region.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Get sample metadata including a mix of seq_type</span></span></span>
<span class="r-in"><span><span class="va">all_types_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>            <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op">==</span> <span class="st">"BL"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">all_types_meta</span>, <span class="va">seq_type</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n<span class="op">=</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 2</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seq_type     n</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> genome     234</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For MYC and SYNCRIP, return CNV and SSM combined status; for MIR17HG, </span></span></span>
<span class="r-in"><span><span class="co"># return only CNV status; for CCND3 return only SSM status</span></span></span>
<span class="r-in"><span><span class="va">genes_and_cn_threshs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  gene_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MYC"</span>, <span class="st">"MIR17HG"</span>, <span class="st">"CCND3"</span>,<span class="st">"ID3"</span>,<span class="st">"DDX3X"</span>, <span class="st">"SYNCRIP"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  cn_thresh<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">genome_cnv_ssm_status</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu">get_cnv_and_ssm_status</span><span class="op">(</span></span></span>
<span class="r-in"><span>                           <span class="va">genes_and_cn_threshs</span>,</span></span>
<span class="r-in"><span>                           genome_build <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                           <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">all_types_meta</span>,<span class="va">seq_type</span><span class="op">==</span><span class="st">"genome"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                           only_cnv <span class="op">=</span> <span class="st">"MIR17HG"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">genome_cnv_ssm_status</span><span class="op">)</span><span class="op">)</span>    </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 234   6</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genome_cnv_ssm_status</span><span class="op">)</span>   </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           MYC MIR17HG CCND3 ID3 DDX3X SYNCRIP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Akata                       1       1     0   1     1       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BL2                         1       1     0   1     1       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BL30                        1       1     0   1     1       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BL41                        1       1     1   1     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BL70                        1       1     1   1     1       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BLGSP-71-06-00001-01A-11D   0       0     1   1     1       0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">genome_cnv_ssm_status</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     MYC MIR17HG   CCND3     ID3   DDX3X SYNCRIP </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     172      38      68     112     118      12 </span>
<span class="r-in"><span>       </span></span>
<span class="r-in"><span><span class="va">all_path_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">all_path_meta</span> <span class="op">=</span> <span class="fu">check_and_clean_metadata</span><span class="op">(</span><span class="va">all_path_meta</span>,duplicate_action<span class="op">=</span><span class="st">"keep_first"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Duplicate rows (keeping first occurrence) for 'sample_id' and 'seq_type' have been dropped.</span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">all_types_meta</span>, <span class="va">seq_type</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n<span class="op">=</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>             </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 2</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seq_type     n</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> genome     234</span>
<span class="r-in"><span><span class="va">all_pathology_status</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu">get_cnv_and_ssm_status</span><span class="op">(</span></span></span>
<span class="r-in"><span>                           <span class="va">genes_and_cn_threshs</span>,</span></span>
<span class="r-in"><span>                           <span class="va">all_path_meta</span>,</span></span>
<span class="r-in"><span>                           genome_build <span class="op">=</span> <span class="st">"grch37"</span>,</span></span>
<span class="r-in"><span>                           only_cnv <span class="op">=</span> <span class="st">"MIR17HG"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">all_pathology_status</span><span class="op">)</span><span class="op">)</span>   </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 448   6</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all_pathology_status</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           MYC MIR17HG CCND3 ID3 DDX3X SYNCRIP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 01-20260T   1       0     0   0     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 02-13135T   1       0     0   0     1       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 02-20170T   0       1     0   0     0       1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 02-22991T   1       0     0   0     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 03-34157T   0       0     0   0     0       0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 04-24029T   0       0     0   0     0       0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">all_pathology_status</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     MYC MIR17HG   CCND3     ID3   DDX3X SYNCRIP </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     114      62      25       5      25     134 </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostia Dreval.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

