<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get Manta SVs — get_manta_sv • GAMBLR.open</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get Manta SVs — get_manta_sv"><meta name="description" content="Retrieve Manta SVs for one or many samples"><meta property="og:description" content="Retrieve Manta SVs for one or many samples"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.open</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-open.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/mutation_summary.html">Exploring simple somatic mutations</a></li>
    <li><a class="dropdown-item" href="../articles/feature_matrix.html">Summarizing mutation status as a feature matrix</a></li>
    <li><a class="dropdown-item" href="../articles/genetic_classification.html">Genetic classification of tumours</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get Manta SVs</h1>

      <div class="d-none name"><code>get_manta_sv.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Retrieve Manta SVs for one or many samples</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_manta_sv</span><span class="op">(</span></span>
<span>  these_samples_metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  <span class="va">region</span>,</span>
<span>  min_vaf <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  min_score <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  pass_filters <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">chromosome</span>,</span>
<span>  <span class="va">qstart</span>,</span>
<span>  <span class="va">qend</span>,</span>
<span>  <span class="va">pairing_status</span>,</span>
<span>  these_sample_ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>A metadata data frame to limit the
result to sample_ids within it</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>The projection genome build. Default is grch37.</p></dd>


<dt id="arg-region">region<a class="anchor" aria-label="anchor" href="#arg-region"></a></dt>
<dd><p>Specify a single region to fetch SVs anchored within
using the format "chrom:start-end"</p></dd>


<dt id="arg-min-vaf">min_vaf<a class="anchor" aria-label="anchor" href="#arg-min-vaf"></a></dt>
<dd><p>The minimum tumour VAF for a SV to be returned.
Default is 0.1.</p></dd>


<dt id="arg-min-score">min_score<a class="anchor" aria-label="anchor" href="#arg-min-score"></a></dt>
<dd><p>The lowest Manta somatic score for a SV to be returned.
Default is 40.</p></dd>


<dt id="arg-pass-filters">pass_filters<a class="anchor" aria-label="anchor" href="#arg-pass-filters"></a></dt>
<dd><p>If TRUE (default) only return SVs that are annotated with
PASS in the FILTER column. Set to FALSE to keep all variants,
regardless if they PASS the filters.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Set to FALSE to minimize the output to console.
Default is TRUE. This parameter also dictates the verbose-ness of
any helper function internally called inside the main function.</p></dd>


<dt id="arg-chromosome">chromosome<a class="anchor" aria-label="anchor" href="#arg-chromosome"></a></dt>
<dd><p>DEPRECATED. Use <code>region</code> instead.</p></dd>


<dt id="arg-qstart">qstart<a class="anchor" aria-label="anchor" href="#arg-qstart"></a></dt>
<dd><p>DEPRECATED. Use <code>region</code> instead.</p></dd>


<dt id="arg-qend">qend<a class="anchor" aria-label="anchor" href="#arg-qend"></a></dt>
<dd><p>DEPRECATED. Use <code>region</code> instead.</p></dd>


<dt id="arg-pairing-status">pairing_status<a class="anchor" aria-label="anchor" href="#arg-pairing-status"></a></dt>
<dd><p>DEPRECATED.</p></dd>


<dt id="arg-these-sample-ids">these_sample_ids<a class="anchor" aria-label="anchor" href="#arg-these-sample-ids"></a></dt>
<dd><p>DEPRECATED.
Subset your metadata and supply `these_samples_metadata“ instead.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Any additional parameters.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Retrieve Manta SVs with additional VCF information to allow for
filtering of high-confidence variants.
To get SV calls for multiple samples, supply a metadata table via
<code>these_samples_metadata</code> that has been subset to only those samples.
The results will be restricted to the sample_ids within that data frame.
This function can also restrict the returned breakpoints within a genomic
region specified via <code>region</code> (in chr:start-end format).
Useful filtering parameters are also available, use <code>min_vaf</code> to set the
minimum tumour VAF for a SV to be returned and <code>min_score</code>
to set the lowest Manta somatic score for a SV to be returned.
In addition, the user can chose to return all variants, even
the ones not passing the filter criteria. To do so,
set <code>pass_filters = FALSE</code> (defaults to TRUE).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># lazily get every SV in the table with default quality filters</span></span></span>
<span class="r-in"><span><span class="va">all_sv</span> <span class="op">&lt;-</span> <span class="fu">get_manta_sv</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled metadata in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled Manta SV (.bedpe) calls in GAMBLR.data...</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM_A   START_A     END_A CHROM_B   START_B     END_B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       1 161658631 161658631       3  16509907  16509907</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       1 161663959 161663959       9  37363320  37363320</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       1 161663959 161663959       9  37363320  37363320</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      11  65267283  65267283      14 106110907 106110907</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      11  65267422  65267422      14 106110905 106110905</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      13  91976545  91976545      14 106211857 106211857</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      manta_name SCORE STRAND_A STRAND_B tumour_sample_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      MantaBND:21171:0:1:0:0:0   133        +        +         FL2002T1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     MantaBND:206628:0:1:0:0:0   122        +        +  09-15842_tumorA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     MantaBND:195941:0:1:0:0:0   151        +        +  09-15842_tumorB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   MantaBND:152220:0:1:0:0:0:0    88        +        -        15-38154T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   MantaBND:152220:0:1:0:0:0:0   135        -        +        15-38154T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 MantaBND:18:59794:59817:0:1:0    90        -        +        15-31924T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   normal_sample_id VAF_tumour  DP pair_status FILTER</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          FL2002N      0.331 127     matched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  09-15842_normal      0.281 196     matched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  09-15842_normal      0.364 187     matched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        15-38154N      0.150 167     matched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        15-38154N      0.290 169     matched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        15-31924N      0.365  85     matched   PASS</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># get all SVs for just one cohort</span></span></span>
<span class="r-in"><span><span class="va">cohort_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="get_gambl_metadata.html">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>              <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cohort</span> <span class="op">==</span> <span class="st">"DLBCL_cell_lines"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">some_sv</span> <span class="op">&lt;-</span> <span class="fu">get_manta_sv</span><span class="op">(</span>these_samples_metadata <span class="op">=</span> <span class="va">cohort_meta</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled Manta SV (.bedpe) calls in GAMBLR.data...</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">some_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM_A   START_A     END_A CHROM_B  START_B    END_B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      14 106329462 106329462      18 60774579 60774579</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      14 106329465 106329465      18 60793497 60793497</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      14 106330466 106330466      18 60793914 60793914</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      14 106349765 106349765      18 60793914 60793914</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      14 106379091 106379091      18 60793492 60793492</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      14 106380227 106380227      18 60774578 60774578</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  manta_name SCORE STRAND_A STRAND_B tumour_sample_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 MantaBND:220769:1:2:0:0:0   134        +        -        SU-DHL-10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 MantaBND:194451:1:2:0:0:0   103        +        -           DOHH-2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 MantaBND:217561:1:2:0:0:0   182        +        -         SU-DHL-4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 MantaBND:217561:0:1:0:0:0   198        -        +         SU-DHL-4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 MantaBND:194451:0:1:0:0:0    91        -        +           DOHH-2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 MantaBND:220769:0:1:0:0:0   169        -        +        SU-DHL-10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   normal_sample_id VAF_tumour DP pair_status FILTER</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        14-11247N      0.318 66   unmatched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        14-11247N      0.290 69   unmatched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        14-11247N      0.474 57   unmatched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        14-11247N      0.500 62   unmatched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        14-11247N      0.300 60   unmatched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6        14-11247N      0.578 45   unmatched   PASS</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">some_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 13</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># get the SVs in a region around MYC</span></span></span>
<span class="r-in"><span><span class="co"># WARNING: This is not the best way to find MYC SVs.</span></span></span>
<span class="r-in"><span><span class="co"># Use annotate_sv on the full SV set instead.</span></span></span>
<span class="r-in"><span><span class="va">myc_region_hg38</span> <span class="op">=</span> <span class="st">"chr8:127710883-127761821"</span></span></span>
<span class="r-in"><span><span class="va">myc_region_grch37</span> <span class="op">=</span> <span class="st">"8:128723128-128774067"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">hg38_myc_locus_sv</span> <span class="op">&lt;-</span> <span class="fu">get_manta_sv</span><span class="op">(</span>region <span class="op">=</span> <span class="va">myc_region_hg38</span>,</span></span>
<span class="r-in"><span>                                projection <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                                verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled metadata in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled Manta SV (.bedpe) calls in GAMBLR.data...</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hg38_myc_locus_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: hg38 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CHROM_A   START_A     END_A CHROM_B   START_B     END_B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    chr2  88860304  88860306    chr8 127751936 127751938</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    chr2  88860417  88860417    chr8 127751955 127751955</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    chr2  88861500  88861500    chr8 127748752 127748752</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    chr3 187811601 187811601    chr8 127745649 127745649</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    chr8 127741233 127741234   chr12  25049104  25049105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    chr8 127713694 127713694   chr14 105857950 105857950</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     manta_name SCORE STRAND_A STRAND_B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  MantaBND:194837:0:1:0:0:0:0   102        +        +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  MantaBND:194837:0:1:0:0:0:0    73        -        -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   MantaBND:1102030:0:1:0:0:0    89        +        +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     MantaBND:48510:0:2:0:0:0   106        -        +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    MantaBND:174836:0:1:0:0:0   219        +        +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 MantaBND:1:10030:23823:0:0:0   109        -        +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            tumour_sample_id          normal_sample_id VAF_tumour  DP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 BLGSP-71-27-00414-01A-01E BLGSP-71-27-00414-10A-01D      0.171 280</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 BLGSP-71-27-00414-01A-01E BLGSP-71-27-00414-10A-01D      0.117 230</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 BLGSP-71-30-00647-01A-01E BLGSP-71-06-00286-99A-01D      0.283  46</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  FL1008T2                   FL1008N      0.171 245</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                  FL1018T2                   FL1018N      0.323 288</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 BLGSP-71-06-00280-01A-01D BLGSP-71-06-00280-99A-01D      0.272 235</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   pair_status FILTER</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     matched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     matched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   unmatched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     matched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5     matched   PASS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6     matched   PASS</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">hg38_myc_locus_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 260</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">incorrect_myc_locus_sv</span> <span class="op">&lt;-</span> <span class="fu">get_manta_sv</span><span class="op">(</span>region <span class="op">=</span> <span class="va">myc_region_grch37</span>,</span></span>
<span class="r-in"><span>                                projection <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                                verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled metadata in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled Manta SV (.bedpe) calls in GAMBLR.data...</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">incorrect_myc_locus_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: hg38 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] CHROM_A          START_A          END_A            CHROM_B         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] START_B          END_B            manta_name       SCORE           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] STRAND_A         STRAND_B         tumour_sample_id normal_sample_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] VAF_tumour       DP               pair_status      FILTER          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;0 rows&gt; (or 0-length row.names)</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">incorrect_myc_locus_sv</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-in"><span><span class="co"># The effect of specifying the wrong coordinate is evident</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Despite potentially being incomplete, we can nonetheless</span></span></span>
<span class="r-in"><span><span class="co"># annotate these directly for more details</span></span></span>
<span class="r-in"><span><span class="va">annotated_myc_hg38</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>         <span class="fu">annotate_sv</span><span class="op">(</span><span class="va">hg38_myc_locus_sv</span>, genome_build <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">annotated_myc_hg38</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chrom1    start1      end1 chrom2    start2      end2 name score strand1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      8 127741233 127741234     12  25049104  25049105    .   219       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      8 127713694 127713694     14 105857950 105857950    .   109       -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      8 127716025 127716934     14 105862581 105863164    .   112       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      8 127716523 127716523     14 105862757 105862757    .   173       -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      8 127718148 127718148     14 105860256 105860256    .   152       +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      8 127718150 127718150     14 105860564 105860564    .   163       -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   strand2          tumour_sample_id gene partner   fusion</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1       +                  FL1018T2  MYC    LRMP LRMP-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       + BLGSP-71-06-00280-01A-01D  MYC     IGH  IGH-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       - BLGSP-71-06-00084-01A-01D  MYC     IGH  IGH-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       + BLGSP-71-06-00084-01A-01D  MYC     IGH  IGH-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       - BLGSP-71-08-00036-01A-01D  MYC     IGH  IGH-MYC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       + BLGSP-71-08-00036-01A-01D  MYC     IGH  IGH-MYC</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">annotated_myc_hg38</span><span class="op">$</span><span class="va">partner</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BCL6  DMD  IGH  IGK  IGL LRMP PAX5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    1    2  247    3    2    1    4 </span>
<span class="r-in"><span><span class="co"># The usual MYC partners are seen here</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostia Dreval.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

