<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get GAMBL Metadata. — get_gambl_metadata • GAMBLR.open</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get GAMBL Metadata. — get_gambl_metadata"><meta name="description" content="Convenience function for loading the sample metadata."><meta property="og:description" content="Convenience function for loading the sample metadata."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.open</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-open.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/mutation_summary.html">Exploring simple somatic mutations</a></li>
    <li><a class="dropdown-item" href="../articles/feature_matrix.html">Summarizing mutation status as a feature matrix</a></li>
    <li><a class="dropdown-item" href="../articles/genetic_classification.html">Genetic classification of tumours</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get GAMBL Metadata.</h1>

      <div class="d-none name"><code>get_gambl_metadata.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Convenience function for loading the sample metadata.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_gambl_metadata</span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genome"</span>, <span class="st">"capture"</span><span class="op">)</span>, <span class="va">case_set</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-seq-type-filter">seq_type_filter<a class="anchor" aria-label="anchor" href="#arg-seq-type-filter"></a></dt>
<dd><p>Specify the seq type you want to return metadata for.
Default is "genome".</p></dd>


<dt id="arg-case-set">case_set<a class="anchor" aria-label="anchor" href="#arg-case-set"></a></dt>
<dd><p>Optionally specify study details to return samples from a
particular case set. See function description for supported case sets.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Any additional parameters.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with metadata, tailored for user without GSC access.</p>
<dl><dt>compression</dt>
<dd><p>Format of the original data used as input for our analysis pipelines (cram, bam or fastq)</p></dd>

<dt>bam_available</dt>
<dd><p>Whether or not this file was available when last checked.</p></dd>

<dt>patient_id</dt>
<dd><p>The anonymized unique identifier for this patient. For BC samples, this will be Res ID.</p></dd>

<dt>sample_id</dt>
<dd><p>A unique identifier for the sample analyzed.</p></dd>

<dt>seq_type</dt>
<dd><p>The assay type used to produce this data (one of "genome","capture, "mrna", "promethION")</p></dd>

<dt>genome_build</dt>
<dd><p>The name of the genome reference the data were aligned to.</p></dd>

<dt>cohort</dt>
<dd><p>Name for a group of samples that were added together (usually from a single study), often in the format pathology_cohort_descriptor.</p></dd>

<dt>pathology</dt>
<dd><p>The diagnosis or pathology for the sample</p></dd>

<dt>time_point</dt>
<dd><p>Timing of biopsy in increasing alphabetical order (A = diagnosis, B = first relapse etc)</p></dd>

<dt>ffpe_or_frozen</dt>
<dd><p>Whether the nucleic acids were extracted from a frozen or FFPE sample</p></dd>

<dt>COO_consensus</dt>
<dd><p>Consensus call of COO between different sources.</p></dd>

<dt>DHITsig_consensus</dt>
<dd><p>Consensus call of DHIT signature status between different sources.</p></dd>

<dt>EBV_status_inf</dt>
<dd><p>Inferred EBV status of the tumor</p></dd>

<dt>lymphgen_no_cnv</dt>
<dd><p>LymphGen label using model without CNV</p></dd>

<dt>lymphgen_with_cnv</dt>
<dd><p>LymphGen label using model with CNV</p></dd>

<dt>lymphgen_cnv_noA53</dt>
<dd><p>LymphGen label using model with CNV but excluding A53 class</p></dd>

<dt>lymphgen_wright</dt>
<dd><p>The LymphGen call for this sample from Wright et all (if applicable)</p></dd>

<dt>fl_grade</dt>
<dd><p>Grade of FL samples</p></dd>

<dt>normal_sample_id</dt>
<dd><p>Sample id for normal tissue used in the analysis</p></dd>

<dt>pairing_status</dt>
<dd><p>Matching status of the sample</p></dd>

<dt>lymphgen</dt>
<dd><p>LymphGen label</p></dd>

<dt>molecular_BL</dt>
<dd><p>label of the sample according to the molecular BL classifier</p></dd>

<dt>Tumor_Sample_Barcode</dt>
<dd><p>Duplicate of sample_id for simplifying joins to MAF data frames</p></dd>

<dt>pathology_rank</dt>
<dd><p>Numeric rank for consistent ordering of samples by pathology</p></dd>

<dt>hiv_status</dt>
<dd><p>HIV status of the sample</p></dd>

<dt>age_group</dt>
<dd><p>Adult_BL or Pediatric_BL or Other, specific to the BLGSP study</p></dd>

<dt>sex</dt>
<dd><p>The biological sex of the patient, if available. Allowable options: M, F, NA</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This bare bones function was developed to retrieve metadata for
non-GSC-users. Specify the seq type (<code>seq_type_filter</code>) for the samples you
want returned as the only argument.
It relies on the bundled metadata in this package.
Specify <code>case_set</code> argument to retreive samples from particular study.
Currently supported case_sets are: FL_Dreval (FL samples from Dreval et al),
DLBCL_Dreval (DLBCL samples from Dreval et al), FL-DLBCL-study (all samples
from Dreval et al), DLBCL_Arthur (all samples from Arthur et al study),
DLBCL_Hilton (all samples from Hilton et al DLBCL Trios study),
DLBCL_cell_lines (5 DLBCL cell lines), DLBCL_Chapuy (all samples from Chapuy
et al study), DLBCL_Schmitz (all samples from Schmitz et al study),
DLBCL_Reddy (all samples from Reddy et al study), DLBCL_Thomas (HTMCP DLBCLs
from Thomas et al study), BL_Thomas (BL samples from Thomas et al study)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#return metadata for genome samples (here, the parameter is redundant because 'genome' is the default)</span></span></span>
<span class="r-in"><span><span class="va">genome_meta</span> <span class="op">=</span> <span class="fu">get_gambl_metadata</span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled metadata in GAMBLR.data...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#return metadata for capture samples.</span></span></span>
<span class="r-in"><span><span class="va">capture_meta</span> <span class="op">=</span> <span class="fu">get_gambl_metadata</span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"capture"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled metadata in GAMBLR.data...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#If you want metadata for genome and capture samples you can provide a vector of seq types</span></span></span>
<span class="r-in"><span><span class="va">all_meta</span> <span class="op">=</span> <span class="fu">get_gambl_metadata</span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genome"</span>, <span class="st">"capture"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled metadata in GAMBLR.data...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">all_meta</span>,<span class="va">cohort</span>,<span class="va">seq_type</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 21 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Groups:   cohort, seq_type [21]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    cohort             seq_type     n</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> BL_Adult           genome      91</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> BL_Pediatric       genome     121</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> BL_cell_lines      genome      22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> DLBCL_Gascoyne     genome      21</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> DLBCL_GenomeCanada genome      59</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> DLBCL_HTMCP        genome      43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> DLBCL_ICGC         genome      84</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> DLBCL_LSARP_Trios  capture     12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> DLBCL_LSARP_Trios  genome     142</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> DLBCL_Marra        genome      38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 11 more rows</span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostia Dreval.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

