<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ GAMBLR.predict</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.predict</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/index.html">DLBCLone: Overview</a></li>
    <li><a class="dropdown-item" href="../articles/umap.html">Exploring and selecting features</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_optimize_params.html">Training a new DLBCLone model</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_stored.html">DLBCLone Models</a></li>
    <li><a class="dropdown-item" href="../articles/predict_samples.html">Classifying samples with DLBCLone</a></li>
    <li><a class="dropdown-item" href="../articles/validation.html">Applying a DLBCLone model to new data</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_KNN.html">K-Nearest Neighbors without UMAP</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_mixture_model.html">Gaussian Mixture Model</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="dlbclone-data-pre-processing">DLBCLone: Data pre-processing<a class="anchor" aria-label="anchor" href="#dlbclone-data-pre-processing"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="assemble_genetic_features.html">assemble_genetic_features()</a></code>

        </dt>
        <dd>Assemble genetic features for UMAP input</dd>
      </dl></div><div class="section level2">
      <h2 id="dlbclone-umap">DLBCLone: UMAP<a class="anchor" aria-label="anchor" href="#dlbclone-umap"></a></h2>

      <div class="section-desc"><p>Functions for generating and annotating UMAP visualizations.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="make_and_annotate_umap.html">make_and_annotate_umap()</a></code>

        </dt>
        <dd>Run UMAP and attach result to metadata</dd>
      </dl><dl><dt>

          <code><a href="make_umap_scatterplot.html">make_umap_scatterplot()</a></code>

        </dt>
        <dd>Make UMAP scatterplot</dd>
      </dl><dl><dt>

          <code><a href="basic_umap_scatterplot.html">basic_umap_scatterplot()</a></code>

        </dt>
        <dd>Basic UMAP Scatterplot</dd>
      </dl></div><div class="section level2">
      <h2 id="dlbclone-functions-optimizing-parameters">DLBCLone Functions - Optimizing Parameters<a class="anchor" aria-label="anchor" href="#dlbclone-functions-optimizing-parameters"></a></h2>

      <div class="section-desc"><p>Tools for feature selection, model optimization and QC.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="DLBCLone_optimize_params.html">DLBCLone_optimize_params()</a></code>

        </dt>
        <dd>Optimize parameters for classifying samples using UMAP and k-nearest neighbor</dd>
      </dl><dl><dt>

          <code><a href="posthoc_feature_enrichment.html">posthoc_feature_enrichment()</a></code>

        </dt>
        <dd>Determine feature enrichment per class using truth or predicted labels</dd>
      </dl><dl><dt>

          <code><a href="make_alluvial.html">make_alluvial()</a></code>

        </dt>
        <dd>Create an Alluvial Plot Comparing Original and Predicted Classifications</dd>
      </dl><dl><dt>

          <code><a href="DLBCLone_summarize_model.html">DLBCLone_summarize_model()</a></code>

        </dt>
        <dd>Summarize and Export DLBCLone Model Results</dd>
      </dl><dl><dt>

          <code><a href="DLBCLone_ensemble_postprocess.html">DLBCLone_ensemble_postprocess()</a></code>

        </dt>
        <dd>Post-process KNN results across K to score consistency, (optionally) refine classified/Other cutoffs per-class and (optionally) assign composite classes</dd>
      </dl></div><div class="section level2">
      <h2 id="dlbclone-model-persistence">DLBCLone: Model persistence<a class="anchor" aria-label="anchor" href="#dlbclone-model-persistence"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="DLBCLone_save_optimized.html">DLBCLone_save_optimized()</a></code>

        </dt>
        <dd>Save a DLBCLone model (and optionally integrity test embeddings)</dd>
      </dl><dl><dt>

          <code><a href="DLBCLone_load_optimized.html">DLBCLone_load_optimized()</a></code>

        </dt>
        <dd>Load a previously saved DLBCLone model (including UMAP state)</dd>
      </dl><dl><dt>

          <code><a href="DLBCLone_activate.html">DLBCLone_activate()</a></code>

        </dt>
        <dd>Activate a DLBCLone model by embedding its training set once</dd>
      </dl></div><div class="section level2">
      <h2 id="dlbclone-classifying-samples">DLBCLone: Classifying samples<a class="anchor" aria-label="anchor" href="#dlbclone-classifying-samples"></a></h2>

      <div class="section-desc"><p>Functions for applying trained DLBCLone models to new samples and generating visual summaries of prediction confidence and neighborhood relationships.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="DLBCLone_predict.html">DLBCLone_predict()</a></code>

        </dt>
        <dd>Predict DLBCL genetic subgroup for one or more samples using a pre-trained DLBCLone model</dd>
      </dl><dl><dt>

          <code><a href="make_neighborhood_plot.html">make_neighborhood_plot()</a></code>

        </dt>
        <dd>Make Neighborhood Plot</dd>
      </dl><dl><dt>

          <code><a href="nearest_neighbor_heatmap.html">nearest_neighbor_heatmap()</a></code>

        </dt>
        <dd>Heatmap visualization of mutations in nearest neighbors for a sample</dd>
      </dl></div><div class="section level2">
      <h2 id="dlbclone-k-nearest-neighbors">DLBCLone: K-Nearest Neighbors<a class="anchor" aria-label="anchor" href="#dlbclone-k-nearest-neighbors"></a></h2>

      <div class="section-desc"><p>Functions to train and predict using KNN in high-dimensional space instead of UMAP. Not part of the core DLBCLone functionality.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="DLBCLone_KNN.html">DLBCLone_KNN()</a></code>

        </dt>
        <dd>Run DLBCLone KNN Classification</dd>
      </dl><dl><dt>

          <code><a href="DLBCLone_KNN_predict.html">DLBCLone_KNN_predict()</a></code>

        </dt>
        <dd>Predict DLBCLone Classes for New Samples Using a Trained KNN Model</dd>
      </dl></div><div class="section level2">
      <h2 id="dlbclone-gaussian-mixture-model">DLBCLone: Gaussian Mixture Model<a class="anchor" aria-label="anchor" href="#dlbclone-gaussian-mixture-model"></a></h2>

      <div class="section-desc"><p>Functions to train and predict using gaussian mixture models in UMAP space.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="DLBCLone_train_mixture_model.html">DLBCLone_train_mixture_model()</a></code>

        </dt>
        <dd>Train a Gaussian Mixture Model for DLBCLone Classification</dd>
      </dl><dl><dt>

          <code><a href="DLBCLone_predict_mixture_model.html">DLBCLone_predict_mixture_model()</a></code>

        </dt>
        <dd>Predict DLBCLone Class Membership Using a Trained Gaussian Mixture Model</dd>
      </dl></div><div class="section level2">
      <h2 id="other">Other<a class="anchor" aria-label="anchor" href="#other"></a></h2>

      <div class="section-desc"><p>All remaining GAMBLR.predict functions not listed above.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="DLBCLone_shiny.html">DLBCLone_shiny()</a></code>

        </dt>
        <dd>Run the DLBCLone Shiny App</dd>
      </dl><dl><dt>

          <code><a href="DLBCLone_train_test_plot.html">DLBCLone_train_test_plot()</a></code>

        </dt>
        <dd>Plot the result of a DLBCLone classification</dd>
      </dl><dl><dt>

          <code><a href="RFmodel_BL.html">RFmodel_BL</a></code>

        </dt>
        <dd>BL Classifier model.</dd>
      </dl><dl><dt>

          <code><a href="RFmodel_FL.html">RFmodel_FL</a></code>

        </dt>
        <dd>FL Classifier model.</dd>
      </dl><dl><dt>

          <code><a href="RFmodel_Lacy.html">RFmodel_Lacy</a></code>

        </dt>
        <dd>DLBCL Classifier model.</dd>
      </dl><dl><dt>

          <code><a href="chapuy_features.html">chapuy_features</a></code>

        </dt>
        <dd>Features for DLBCL grouping by Chapuy method.</dd>
      </dl><dl><dt>

          <code><a href="check_for_missing_features.html">check_for_missing_features()</a></code>

        </dt>
        <dd>Check matrix against missing features.</dd>
      </dl><dl><dt>

          <code><a href="classify_bl.html">classify_bl()</a></code>

        </dt>
        <dd>Classify BL samples into genetic subgroups.</dd>
      </dl><dl><dt>

          <code><a href="classify_dlbcl.html">classify_dlbcl()</a></code>

        </dt>
        <dd>Classify DLBCLs according to genetic subgroups.</dd>
      </dl><dl><dt>

          <code><a href="classify_dlbcl_chapuy.html">classify_dlbcl_chapuy()</a></code>

        </dt>
        <dd>Classify DLBCLs according to genetic subgroups of Chapuy et al.</dd>
      </dl><dl><dt>

          <code><a href="classify_dlbcl_lacy.html">classify_dlbcl_lacy()</a></code>

        </dt>
        <dd>Classify DLBCLs according to genetic subgroups of Lacy et al.</dd>
      </dl><dl><dt>

          <code><a href="classify_dlbcl_lymphgenerator.html">classify_dlbcl_lymphgenerator()</a></code>

        </dt>
        <dd>Construct LymphGenerator matrix</dd>
      </dl><dl><dt>

          <code><a href="classify_fl.html">classify_fl()</a></code>

        </dt>
        <dd>Classify FL samples into cFL/dFL subgroups.</dd>
      </dl><dl><dt>

          <code><a href="complete_missing_from_matrix.html">complete_missing_from_matrix()</a></code>

        </dt>
        <dd>Complete samples missing from matrix.</dd>
      </dl><dl><dt>

          <code><a href="construct_reduced_winning_version.html">construct_reduced_winning_version()</a></code>

        </dt>
        <dd>Construct reduced 21-dimension feature vector for DLBCLass</dd>
      </dl><dl><dt>

          <code><a href="flatten_feature.html">flatten_feature()</a></code>

        </dt>
        <dd>Flatten feature</dd>
      </dl><dl><dt>

          <code><a href="handle_genome_build.html">handle_genome_build()</a></code>

        </dt>
        <dd>Harmonize different flavors of genome builds.</dd>
      </dl><dl><dt>

          <code><a href="lacy_features.html">lacy_features</a></code>

        </dt>
        <dd>Features for DLBCL grouping by Lacy method.</dd>
      </dl><dl><dt>

          <code><a href="lymphgenerator_features.html">lymphgenerator_features</a></code>

        </dt>
        <dd>Features for DLBCL grouping by unified method.</dd>
      </dl><dl><dt>

          <code><a href="massage_matrix_for_clustering.html">massage_matrix_for_clustering()</a></code>

        </dt>
        <dd>Will prepare the data frame of binary matrix to be used as NMF input. This means that for the features with SSM and CNV, they will be squished together as one feature named GeneName-MUTorAMP or GeneName-MUTorLOSS, so the CNV features in the input data frame are expected to be named GeneName_AMP or GeneName_LOSS. Next, for the genes with hotspot mutations labelled in the input data as GeneNameHOTSPOT, the feature for hotspot mutation will be given preference and SSM with/without CNV will be set to 0 for that sample. The naming scheme of the features as in this description is important, because the function uses regex to searh for these patters as specified. Finally, if any features are provided to be dropped explicitly, they will be removed, and then the features not meeting the specified minimal frequency will be removed, as well as any samples with 0 features. Consistent with NMF input, in the input data frame each row is a feature, and each column is a sample. The input is expected to be numeric 1/0 with row and column names.</dd>
      </dl><dl><dt>

          <code><a href="optimize_outgroup.html">optimize_outgroup()</a></code>

        </dt>
        <dd>Optimize the threshold for classifying samples as "Other"</dd>
      </dl><dl><dt>

          <code><a href="optimize_purity.html">optimize_purity()</a></code>

        </dt>
        <dd>Optimize Purity Threshold for Classification Assignment</dd>
      </dl><dl><dt>

          <code><a href="process_votes.html">process_votes()</a></code>

        </dt>
        <dd>Process KNN Vote Strings and Scores for Classification</dd>
      </dl><dl><dt>

          <code><a href="report_accuracy.html">report_accuracy()</a></code>

        </dt>
        <dd>Calculate Classification Accuracy and Per-Class Metrics based on Predictions</dd>
      </dl><dl><dt>

          <code><a href="summarize_all_ssm_status.html">summarize_all_ssm_status()</a></code>

        </dt>
        <dd>Summarize SSM (Somatic Single Nucleotide Mutation) Status Across Samples</dd>
      </dl><dl><dt>

          <code><a href="tabulate_ssm_status.html">tabulate_ssm_status()</a></code>

        </dt>
        <dd>Get Coding SSM Status.</dd>
      </dl><dl><dt>

          <code><a href="weighted_knn_predict_with_conf.html">weighted_knn_predict_with_conf()</a></code>

        </dt>
        <dd>Weighted k-nearest neighbor with confidence estimate</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kostiantyn Dreval, Ryan Morin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

