<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Summarize SSM (Somatic Single Nucleotide Mutation) Status Across Samples — summarize_all_ssm_status • GAMBLR.predict</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Summarize SSM (Somatic Single Nucleotide Mutation) Status Across Samples — summarize_all_ssm_status"><meta name="description" content="This function summarizes the mutation status for a set of genes
across multiple samples, separating mutations by class for genes specified
by separate_by_class_genes and counting the number of hits
in each sample per mutation category."><meta property="og:description" content="This function summarizes the mutation status for a set of genes
across multiple samples, separating mutations by class for genes specified
by separate_by_class_genes and counting the number of hits
in each sample per mutation category."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.predict</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/index.html">DLBCLone: Overview</a></li>
    <li><a class="dropdown-item" href="../articles/umap.html">Exploring and selecting features</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_optimize_params.html">Training a new DLBCLone model</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_stored.html">DLBCLone Models</a></li>
    <li><a class="dropdown-item" href="../articles/predict_samples.html">Classifying samples with DLBCLone</a></li>
    <li><a class="dropdown-item" href="../articles/full_walkthrough.html">DLBCLone: Full Walkthrough</a></li>
    <li><a class="dropdown-item" href="../articles/validation.html">Applying a DLBCLone model to new data</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_KNN.html">K-Nearest Neighbors without UMAP</a></li>
    <li><a class="dropdown-item" href="../articles/DLBCLone_mixture_model.html">Gaussian Mixture Model</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Summarize SSM (Somatic Single Nucleotide Mutation) Status Across Samples</h1>

      <div class="d-none name"><code>summarize_all_ssm_status.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function summarizes the mutation status for a set of genes
across multiple samples, separating mutations by class for genes specified
by <code>separate_by_class_genes</code> and counting the number of hits
in each sample per mutation category.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">summarize_all_ssm_status</span><span class="op">(</span></span>
<span>  <span class="va">maf_df</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  <span class="va">genes_of_interest</span>,</span>
<span>  <span class="va">synon_genes</span>,</span>
<span>  <span class="va">silent_maf_df</span>,</span>
<span>  separate_by_class_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  count_hits <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-maf-df">maf_df<a class="anchor" aria-label="anchor" href="#arg-maf-df"></a></dt>
<dd><p>A data frame containing mutation annotation format
(MAF) data, with at least the following columns:
<code>Hugo_Symbol</code>, <code>Variant_Classification</code>, and
<code>Tumor_Sample_Barcode</code>.</p></dd>


<dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>A data frame containing metadata for the samples,
with at least a <code>sample_id</code> column.
Any sample that does not have a matching sample_id in these_samples_metadata will be dropped.</p></dd>


<dt id="arg-genes-of-interest">genes_of_interest<a class="anchor" aria-label="anchor" href="#arg-genes-of-interest"></a></dt>
<dd><p>A character vector of gene symbols to include
in the summary. If missing, defaults to all Tier 1 B-cell lymphoma genes.</p></dd>


<dt id="arg-synon-genes">synon_genes<a class="anchor" aria-label="anchor" href="#arg-synon-genes"></a></dt>
<dd><p>(Optional) A character vector of gene symbols for which
synonymous mutations should be included.</p></dd>


<dt id="arg-silent-maf-df">silent_maf_df<a class="anchor" aria-label="anchor" href="#arg-silent-maf-df"></a></dt>
<dd><p>(Optional) A separate data frame containing
silent mutation data if
the user doesn't want to pull silent mutation status from <code>maf_df</code>.
This argument is useful when you want to combine mutations from
the output of get_coding_ssm and get_ssm_by_region or get_ssm_by_gene</p></dd>


<dt id="arg-separate-by-class-genes">separate_by_class_genes<a class="anchor" aria-label="anchor" href="#arg-separate-by-class-genes"></a></dt>
<dd><p>(Optional) A character vector of
gene symbols for which mutations should be separated by class
(e.g., "Nonsense_Mutation", "Missense_Mutation").</p></dd>


<dt id="arg-count-hits">count_hits<a class="anchor" aria-label="anchor" href="#arg-count-hits"></a></dt>
<dd><p>Logical; if <code>TRUE</code>, counts the number of mutations
per gene per sample. If <code>FALSE</code> (default), only presence/absence is recorded.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A wide-format data frame (matrix) with samples as rows and
mutation types as columns. Each cell contains either the count of
mutations (if <code>count_hits = TRUE</code>) or a binary indicator (0/1)
for mutation presence.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>

<ul><li><p>Mutations are grouped and optionally separated by mutation
class for genes specified in <code>separate_by_class_genes</code></p></li>
<li><p>Synonymous mutations can be counted as another separate feature
for genes specified by <code>synon_genes</code> genes.</p></li>
<li><p>The function simplifies mutation annotations and pivots the data
to a wide format suitable for downstream analysis.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># A basic example, using only the output of get_all_coding_ssm</span></span></span>
<span class="r-in"><span><span class="co"># Since the only non-coding class this function handles is Silent,</span></span></span>
<span class="r-in"><span><span class="co"># we will be missing most non-coding types such as Intron, UTR, Flank</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">sample_metadata</span> <span class="op">=</span> <span class="fu">get_sample_metadata</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seq_type</span><span class="op">!=</span> <span class="st">"mrna"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">maf_data</span> <span class="op">=</span> <span class="fu">get_all_coding_ssm</span><span class="op">(</span><span class="va">sample_metadata</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">mutation_matrix</span> <span class="op">&lt;-</span> <span class="fu">summarize_all_ssm_status</span><span class="op">(</span></span></span>
<span class="r-in"><span>  maf_df <span class="op">=</span> <span class="va">maf_data</span>,</span></span>
<span class="r-in"><span>  these_samples_metadata <span class="op">=</span> <span class="va">sample_metadata</span>,</span></span>
<span class="r-in"><span>  genes_of_interest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"SGK1"</span>, <span class="st">"BCL2"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  synon_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCL2"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  separate_by_class_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>,<span class="st">"SGK1"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  count_hits <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kostiantyn Dreval, Ryan Morin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

